{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6f42b37c",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "Install in R:\n",
    "```\n",
    "install.packages(\"BiocManager\")\n",
    "BiocManager::install(c(\"edgeR\", \"BiocParallel\", \"RhpcBLASctl\"))\n",
    "```\n",
    "\n",
    "Install\n",
    "\n",
    "```pip install toytree ete4 pybiomart```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f73cc46",
   "metadata": {},
   "source": [
    "## Options & Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "099e052f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/easlinger/miniconda3/envs/rsc/lib/python3.13/site-packages/senepy/load_hubs.py:3: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  from pkg_resources import resource_filename\n",
      "/home/easlinger/miniconda3/envs/rsc/lib/python3.13/site-packages/marshmallow/__init__.py:17: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.\n",
      "  __version_info__ = tuple(LooseVersion(__version__).version)\n",
      "/home/easlinger/miniconda3/envs/rsc/lib/python3.13/site-packages/marshmallow/fields.py:198: RemovedInMarshmallow4Warning: Passing field metadata as a keyword arg is deprecated. Use the explicit `metadata=...` argument instead.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import re\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "import logging\n",
    "import seaborn as sns\n",
    "from scipy.sparse import issparse\n",
    "from pymer4.models import lmer, lm, glmer, compare\n",
    "from pymer4 import config\n",
    "import polars\n",
    "import statsmodels.api as sm\n",
    "import senepy\n",
    "import gseapy as gp\n",
    "import pertpy as pt\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scflow\n",
    "from scflow.ax import find_senescence_genes, run_senepy\n",
    "\n",
    "# Display/Environment Settings\n",
    "os.environ[\"KMP_DUPLICATE_LIB_OK\"] = \"True\"\n",
    "warnings.filterwarnings(\"ignore\", message=\".*ast.Ellipsis is deprecated.*\")\n",
    "config.PANDAS_BACKEND = True\n",
    "warnings.filterwarnings(\"ignore\", message=\".*vert.*will be deprecated.*\")\n",
    "pd.set_option(\"display.max_rows\", 500)  # or None for unlimited rows\n",
    "pd.set_option(\"display.max_columns\", 100)\n",
    "pd.set_option(\"display.max_colwidth\", 100)\n",
    "pd.set_option(\"display.width\", 300)\n",
    "warnings.filterwarnings(\n",
    "    \"ignore\", message=\"Bitwise inversion '~' on bool is deprecated\",\n",
    "    category=DeprecationWarning)\n",
    "\n",
    "# Process Options\n",
    "ncpus = os.cpu_count() - 1\n",
    "overwrite = True\n",
    "\n",
    "# If You Want Results Emailed (Will Overwrite HTML Regardless of `overwite`)\n",
    "cur_file = os.path.join(os.path.abspath(\"\"), \"analyze_senescence_celltype\"\n",
    "                        \"_specific_threshold_zeng.ipynb\")\n",
    "html_out = os.path.splitext(cur_file)[0] + \".html\"\n",
    "email = \"elizabeth.aslinger@aya.yale.edu\"\n",
    "# set email to None to skip\n",
    "\n",
    "# Metadata Information\n",
    "col_celltype = \"cluster_name\"\n",
    "species = \"Mouse\"\n",
    "col_sample, col_batch = \"donor_label\", \"Group\"\n",
    "col_condition = \"donor_sex\"\n",
    "col_age = \"donor_age_category\"\n",
    "keys = {col_age: dict(key_control=\"Adult\", key_treatment=\"Aged\"),\n",
    "        col_condition: dict(key_control=\"M\", key_treatment=\"F\")}\n",
    "\n",
    "# Scoring Metrics\n",
    "sen_metrics = [\"senmayo\"]\n",
    "use_metric = \"senmayo\"\n",
    "percentile = 1\n",
    "\n",
    "# File Input & Output\n",
    "file_path = \"data/Zeng/Zeng-Aging-Mouse-10Xv3-raw.h5ad\"\n",
    "file_path_new = os.path.splitext(file_path)[0] + \"_analyzed.h5ad\"\n",
    "\n",
    "# Load Data\n",
    "ann = sc.read(file_path, backed=\"r\")\n",
    "ann.obs = ann.obs.join(pd.read_csv(\n",
    "    \"data/Zeng/cell_metadata.csv\", index_col=0), rsuffix=\"_csv\")\n",
    "# metadata_donor = pd.read_csv(\"data/Zeng/donor.csv\")\n",
    "# metadata_donor[\"donor_label\"] = metadata_donor[\"donor_label\"].astype(\n",
    "#     \"category\")\n",
    "# ann.obs[col_sample] = ann.obs[col_sample].astype(\"category\")\n",
    "# ann.obs = ann.obs.join(metadata_donor, rsuffix=\"_csv\")\n",
    "ann.obs[col_age] = ann.obs[col_age].str.capitalize()\n",
    "ann.obs.loc[:, col_batch] = ann.obs[col_condition] + \" | \" + ann.obs[\n",
    "    col_age].str.capitalize()\n",
    "ann.obs = ann.obs.join(pd.read_csv(\"data/Zeng/cluster.csv\", index_col=0),\n",
    "                       on=\"cluster_alias\", rsuffix=\"_cl_info\")\n",
    "# ann.write(file_path_new)\n",
    "# ann.file.close()\n",
    "# del ann\n",
    "# self = scflow.Rna(ann, col_sample=col_sample, col_celltype=col_celltype,\n",
    "#                   col_batch=col_batch, inplace=True)\n",
    "# self.rna.var_names = list(self.rna.var.gene_symbol)\n",
    "\n",
    "# # Display Object\n",
    "# print(self.rna)\n",
    "# self.rna.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a5213c8",
   "metadata": {},
   "source": [
    "# Get Marker Gene Sets\n",
    "\n",
    "From Google Drive, except senmayo derived from file downloaded from\n",
    "https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/SAUL_SEN_MAYO.html\n",
    "then created using:\n",
    "\n",
    "```\n",
    "from pybiomart import Server\n",
    "\n",
    "senmayo = pd.read_csv(\"resources/SAUL_SEN_MAYO.v2025.1.Hs.tsv\",\n",
    "                      sep=\"\\t\", header=0)\n",
    "senmayo = senmayo[senmayo.STANDARD_NAME == \"GENE_SYMBOLS\"]\n",
    "senmayo = pd.Series({\"Senmayo\": senmayo.iloc[0, 1].split(\n",
    "    \",\")}, name=\"symbol\").rename_axis(\"Gene_Set\").explode().str.strip(\n",
    "        ).replace(\"\", np.nan).dropna().to_frame()\n",
    "server = Server(host=\"http://www.ensembl.org\")\n",
    "mart = server.marts[\"ENSEMBL_MART_ENSEMBL\"] # Access the Ensembl mart\n",
    "human_dataset = mart.datasets[\"hsapiens_gene_ensembl\"]\n",
    "orthologs = human_dataset.query(attributes=[\n",
    "    \"external_gene_name\", \"mmusculus_homolog_associated_gene_name\"])\n",
    "orthologs = orthologs.dropna().set_index(orthologs.columns[0])\n",
    "orthologs = orthologs.loc[orthologs.index.intersection(\n",
    "    senmayo.symbol.to_list())].iloc[:, 0]\n",
    "print(f\"{orthologs.shape[0]} orthologs found out of\"\n",
    "      f\" {len(senmayo.symbol.unique())} senmayo genes\")\n",
    "senmayo = senmayo.assign(symbol=senmayo.replace({\"symbol\": dict(zip(\n",
    "    orthologs.index, orthologs.values))}))\n",
    "senmayo.to_csv(\"gene_sets/senmayo.csv\")\n",
    "```\n",
    "\n",
    "(Also could use `scflow.tl.convert_gene_species(senmayo.symbol.to_list())`)\n",
    "\n",
    "Certain gene sets created from Google Drive ADBR AWG > Projects > White Matter > Gene Sets files (with the suffix = \"_adjusted\") were modified for consistency, e.g.,\n",
    "\n",
    "```\n",
    "pd.read_csv(\"gene_sets/_iron_genes_gmt_out.csv\", index_col=[\n",
    "    0, 1]).rename_axis([\"row_id\", \"pathway\"]).stack().to_frame(\n",
    "        \"symbol\").reset_index(-1, drop=True).reset_index().to_csv(\n",
    "            \"gene_sets/_iron_genes_gmt_out_adjusted.csv\")\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "305ca01b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mks = [pd.read_csv(os.path.join(\"gene_sets\", i)).dropna(\n",
    "    how=\"all\", axis=1).assign(Source_File=i) for i in os.listdir(\"gene_sets\")]\n",
    "mks = [x.assign(Gene_Set=x[\"Source_File\"].iloc[0].split(\"_2025\")[0]) if (\n",
    "    \"pathway\" in x) else x for x in mks]\n",
    "mks = pd.concat(mks).drop(\"Unnamed: 0\", axis=1).drop(\"row_id\", axis=1)\n",
    "mks = mks[mks.symbol.isin(self.rna.var_names)]\n",
    "mks = mks[[\"Gene_Set\", \"symbol\"]].set_index(\"Gene_Set\").groupby(\n",
    "    \"Gene_Set\").apply(lambda x: x[\"symbol\"].to_list())\n",
    "marker_genes_dict = dict(mks)\n",
    "marker_gene_sets = mks.copy()\n",
    "print(marker_gene_sets)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab398745",
   "metadata": {},
   "source": [
    "# Score Senescence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e66bbcb8",
   "metadata": {},
   "source": [
    "## Run Scoring\n",
    "\n",
    "If you want just to find the senescence-related genes, run:\n",
    "\n",
    "```\n",
    "tissue = \"Brain\"  # or whatever you want here\n",
    "celltype = None  # or whatever you want here\n",
    "genes_senepy, hubs, figs = find_senescence_genes(\n",
    "    self.rna, species=species, tissue=tissue, celltype=celltype,\n",
    "    overlap_threshold=0, literature_sources=None,\n",
    "    sengpt_sources=True, plot=True, col_celltype=col_celltype\n",
    ")  # find tissue- and/or cell type-specific genes\n",
    "```\n",
    "\n",
    "To see available hubs:\n",
    "\n",
    "```\n",
    "hubs = senepy.load_hubs(species=species)\n",
    "hubs.metadata.set_index([\"tissue\", \"cell\"]).sort_index()\n",
    "```\n",
    "\n",
    "Other examples of arguments you could use:\n",
    "\n",
    "```\n",
    "# tissue = [\"Brain\", \"Myeloid\"]\n",
    "# tissue = None\n",
    "\n",
    "# celltype = \"GABA interneuron\"  # only human I think\n",
    "# celltype = [\"microglia\", \"t cell\", \"monocyte\"]\n",
    "# celltype = \"microglia\"\n",
    "# celltype = \"microglial cell\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b76ab837",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.score_genes(self.rna, marker_gene_sets.loc[\"Senmayo\"],\n",
    "                  ctrl_as_ref=True, ctrl_size=50, gene_pool=None,\n",
    "                  n_bins=25, score_name=\"score_senmayo\", random_state=0,\n",
    "                  copy=False, layer=\"log1p\", use_raw=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e7d0e58",
   "metadata": {},
   "source": [
    "## Label SnCs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3929ece3",
   "metadata": {},
   "source": [
    "### Define Snc Threshold & Label Cells\n",
    "\n",
    "Based on percentile threshold derived from control group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92c0f016",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine Threshold\n",
    "perc_s, perc = str(100 - percentile) + \"%\", 1 - (percentile / 100)\n",
    "p_h = 3\n",
    "perc_s_h, perc_h = str(100 - p_h) + \"%\", 1 - (p_h / 100)\n",
    "print(\"Percentile for Threshold:\", perc_s)\n",
    "for u in sen_metrics:\n",
    "    self.rna.obs = self.rna.obs.join(self.rna.obs[self.rna.obs[\n",
    "        col_batch] == keys[col_batch][\"key_control\"]].groupby(\n",
    "            col_celltype).apply(lambda x: x[f\"score_{u}\"].describe(\n",
    "                percentiles=[perc])[perc_s], include_groups=False).to_frame(\n",
    "                    f\"Senescence_Threshold_{u}\"), on=col_celltype).loc[\n",
    "                        self.rna.obs.index]\n",
    "    self.rna.obs = self.rna.obs.join(self.rna.obs[self.rna.obs[\n",
    "        col_batch] == keys[col_batch][\"key_control\"]].groupby(\n",
    "            col_celltype).apply(lambda x: x[f\"score_{u}\"].describe(\n",
    "                percentiles=[perc_h])[perc_s_h],\n",
    "                                include_groups=False).to_frame(\n",
    "                                    f\"Senescence_Threshold_{u}_{p_h}\"),\n",
    "                                on=col_celltype).loc[self.rna.obs.index]\n",
    "\n",
    "# Binary CLassification\n",
    "for x in sen_metrics:\n",
    "    self.rna.obs.loc[:, f\"Senescent_Cell_{x}\"] = (self.rna.obs[\n",
    "        f\"score_{x}\"] >= self.rna.obs[f\"Senescence_Threshold_{x}\"])\n",
    "    self.rna.obs.loc[:, f\"Senescent_Cell_{x}_{p_h}\"] = (self.rna.obs[\n",
    "        f\"score_{x}\"] >= self.rna.obs[f\"Senescence_Threshold_{x}_{p_h}\"])\n",
    "\n",
    "# String Labels (for Plotting/Categorical Analysis)\n",
    "for x in sen_metrics:\n",
    "    self.rna.obs.loc[:, f\"Senescent_Cell_Label_{x}\"] = self.rna.obs[\n",
    "        f\"Senescent_Cell_{x}\"].astype(bool).astype(str).replace({\n",
    "            \"True\": \"Senescent\", \"False\": \"Non-Senescent\"})\n",
    "    self.rna.obs.loc[:, f\"Senescent_Cell_Label_hierarchy_{x}\"] = self.rna.obs[\n",
    "        f\"Senescent_Cell_{x}\"].astype(bool).astype(str).replace({\n",
    "            \"True\": \"Senescent\", \"False\": \"Non-Senescent\"})\n",
    "    self.rna.obs.loc[:, f\"Senescent_Cell_Label_{x}_{p_h}\"] = self.rna.obs[\n",
    "        f\"Senescent_Cell_{x}_{p_h}\"].astype(bool).astype(str).replace({\n",
    "            \"True\": \"Senescent\", \"False\": \"Non-Senescent\"})\n",
    "    self.rna.obs.loc[:, f\"Senescent_Cell_Label_by_Type_{x}\"] = self.rna.obs[\n",
    "        f\"Senescent_Cell_Label_{x}\"].replace({\n",
    "            \"Senescent\": \"SnC \", \"Non-Senescent\": \"\"}) + self.rna.obs[\n",
    "                col_celltype].astype(str)\n",
    "    self.rna.obs.loc[:, f\"SnC_hierarchy_{x}\"] = self.rna.obs[\n",
    "        f\"Senescent_Cell_Label_{x}\"].replace({\n",
    "            \"Senescent\": \"SnC \", \"Non-Senescent\": \"\"}) + self.rna.obs[\n",
    "                col_celltype + \"_hierarchy\"].astype(str)\n",
    "\n",
    "# Descriptives\n",
    "for x in sen_metrics:\n",
    "    print(x, round(self.rna.obs[[\n",
    "        f\"Senescent_Cell_Label_{x}\", col_celltype]].value_counts(\n",
    "            normalize=True).sort_index() * 100, 1).unstack(0))\n",
    "\n",
    "# Choose Metrics for Main Analysis\n",
    "for x in [\"Senescent_Cell_Label\", \"Senescent_Cell\", \"SnC_hierarchy\",\n",
    "            \"Senescence_Threshold\", \"Senescent_Cell_Label_by_Type\"]:\n",
    "    self.rna.obs.loc[:, x] = self.rna.obs[f\"{x}_{use_metric}\"]\n",
    "self.rna.obs.loc[:, \"senscore\"] = self.rna.obs[f\"score_{use_metric}\"]\n",
    "self.rna.obs.loc[\n",
    "    :, f\"Senescent_Cell_Label_by_Type_{use_metric}_{p_h}\"] = self.rna.obs[\n",
    "        f\"Senescent_Cell_Label_{use_metric}_{p_h}\"].replace({\n",
    "            \"Non-Senescent\": \"\", \"Senescent\": \"SnC \"}).astype(\n",
    "                str) + self.rna.obs[\n",
    "                    col_celltype].astype(str)  # lower threshold SnC by type\n",
    "genes = marker_gene_sets.loc[\"Senmayo\"] if (\n",
    "    use_metric == \"senmayo\") else genes_senepy  # main senescence gene list\n",
    "\n",
    "# Avoid Nesting Duplication Issues\n",
    "self.rna.obs.loc[:, \"SnC_hierarchy\"] = self.rna.obs[\n",
    "    \"SnC_hierarchy\"].replace({\n",
    "        \"Neuron\": \"Neurons\", \"SnC Neuron\": \"SnC Neurons\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e04a4644",
   "metadata": {},
   "source": [
    "### Label Batch/Cell-Specific Senescence Percentiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e12c4e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for x in [\"Senescence_Label\", \"Senescence_CellType\"]:\n",
    "#     if x in self.rna.obs:\n",
    "#         self.rna.obs = self.rna.obs.drop(x, axis=1)\n",
    "\n",
    "# # %ile Thresholds (Overall)\n",
    "# thresh = self.rna.obs[\"score_senepy\"].describe(\n",
    "#     percentiles=[0.90, 0.95, 0.99])\n",
    "# thres_ct =  self.rna.obs[[\"score_senepy\", col_celltype]].groupby(\n",
    "#     col_celltype).describe(percentiles=[0.90, 0.95, 0.99])[\"score_senepy\"]\n",
    "\n",
    "# sen = self.rna.obs[\"score_senepy\"].apply(\n",
    "#     lambda x: \"Top 1%\" if x >= float(thresh.loc[\"99%\"]) else \"Top 5%\" if (\n",
    "#         x >= float(thresh.loc[\"95%\"])) else \"Top 10%\" if x >= float(\n",
    "#             thresh.loc[\"90%\"]) else \"Bottom 90%\").to_frame(\n",
    "#                 \"Senescence_Label\").loc[self.rna.obs.index]\n",
    "# self.rna.obs = self.rna.obs.join(sen).loc[self.rna.obs.index]\n",
    "\n",
    "# # %ile Thresholds (by Cell Type)\n",
    "# sct = self.rna.obs.groupby(\n",
    "#     col_celltype).apply(lambda g: g[\"score_senepy\"].apply(\n",
    "#         lambda x: f\"Top 1% {g.name}\" if x >= float(thres_ct.loc[g.name][\n",
    "#             \"99%\"]) else f\"Top 5% {g.name}\" if (x >= float(thres_ct.loc[\n",
    "#                 g.name][\"95%\"])) else f\"Top 10% {g.name}\" if x >= float(\n",
    "#                     thres_ct.loc[g.name][\"90%\"]) else g.name),\n",
    "#                         include_groups=False)\n",
    "\n",
    "# sct = sct.to_frame(\"Senescence_CellType\").reset_index(0, drop=True)\n",
    "# sct = sct.loc[self.rna.obs.index]\n",
    "# self.rna.obs = self.rna.obs.join(sct).loc[self.rna.obs.index]\n",
    "\n",
    "# for x in [\"Senescence_Label\", \"Senescence_CellType\"]:\n",
    "#     self.rna.obs = self.rna.obs.assign(**{\n",
    "#         x: self.rna.obs[x].astype(\"category\")})\n",
    "\n",
    "# high_sen_cts = [i for i in self.rna.obs[\"Senescence_CellType\"].unique(\n",
    "#     ) if \" 1% \" in i]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2810c6bb",
   "metadata": {},
   "source": [
    "### Descriptives of SnCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e81ecd71",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Percent Senescent: \\n\\n\", round(self.rna.obs.groupby([\n",
    "    col_celltype, col_batch]).apply(lambda x: x[\"Senescent_Cell\"].astype(\n",
    "        int).value_counts(normalize=True), include_groups=False) * 100,\n",
    "    2).unstack(col_celltype))\n",
    "sns.catplot(self.rna.obs, y=\"Senescent_Cell\", x=col_celltype,\n",
    "            hue=col_batch, kind=\"bar\", height=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d20aec23",
   "metadata": {},
   "source": [
    "# SnC DEGs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88962fe9",
   "metadata": {},
   "source": [
    "## Overall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "912e9a03",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_threshold = 1e-10\n",
    "lfc_threshold = 1\n",
    "self.find_markers(col_celltype=\"Senescent_Cell_Label_by_Type\")\n",
    "df_snc_markers = {}\n",
    "df_snc_markers[\"All\"] = self.get_markers_df(\n",
    "    n_genes=50, col_celltype=\"Senescent_Cell_Label_by_Type\",\n",
    "    log2fc_threshold=1, p_threshold=1e-10)\n",
    "# print(df_snc_markers[\"All\"].loc[pd.unique([\n",
    "#     i[0] for i in df_snc_markers[\"All\"].index.values if (\n",
    "#         \"SnC\" in i[0])])].groupby(\"Senescent_Cell_Label_by_Type\").apply(\n",
    "#             lambda x: x.head(3)))\n",
    "for x in self.rna.obs[col_celltype].unique():\n",
    "    df_snc_markers[x] = self.find_markers(\n",
    "        col_celltype=\"Senescent_Cell_Label_by_Type\",\n",
    "        reference=x, groups=[x, f\"SnC {x}\"], rankby_abs=True,\n",
    "        inplace=False).reset_index(0, drop=True).assign(**{\n",
    "            col_celltype: f\"SnC {x}\"}).reset_index().set_index([\n",
    "                col_celltype, \"names\"])\n",
    "df_snc_markers = pd.concat(df_snc_markers, names=[\"Comparison\"])\n",
    "df_snc_markers = df_snc_markers.rename_axis([\n",
    "    \"Comparison\", col_celltype, \"names\"]).rename({\n",
    "        \"logfoldchanges\": \"log_fc\"}, axis=1).assign(\n",
    "            log_fc_abs=df_snc_markers[\"logfoldchanges\"].abs())\n",
    "df_snc_markers_top = df_snc_markers[df_snc_markers.pvals_adj < p_threshold]\n",
    "df_snc_markers_top = df_snc_markers_top[\n",
    "    df_snc_markers.log_fc_abs >= lfc_threshold]\n",
    "df_snc_markers_top = df_snc_markers_top.groupby([\n",
    "    \"Comparison\", col_celltype]).apply(\n",
    "        lambda x: x.assign(scores_abs=x.scores.abs()).sort_values(\n",
    "            \"scores_abs\", ascending=False).head(20),\n",
    "        include_groups=False).reset_index([0, 1], drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28d39a17",
   "metadata": {},
   "source": [
    "## By Group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b60a96f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_rank_genes, p_threshold = {}, 1e-10\n",
    "for x in self.rna.obs[col_batch].unique():\n",
    "    res_rank_genes[x] = {}\n",
    "    for c in self.rna.obs[self.rna.obs[col_batch] == x][\n",
    "            col_celltype].unique():\n",
    "        s_iter = (self.rna.obs[col_batch] == x) & (self.rna.obs[\n",
    "            \"Senescent_Cell_Label_by_Type\"].isin([c, f\"SnC {c}\"]))\n",
    "        tmp = sc.tl.rank_genes_groups(\n",
    "            self.rna[s_iter], \"Senescent_Cell_Label_by_Type\", rankby_abs=True,\n",
    "            layer=\"log1p\", groups=[c, f\"SnC {c}\"], reference=c, copy=True)\n",
    "        res_rank_genes[x][c] = sc.get.rank_genes_groups_df(\n",
    "            tmp, f\"SnC {c}\", pval_cutoff=p_threshold)\n",
    "    tmp = sc.tl.rank_genes_groups(\n",
    "            self.rna[self.rna.obs[col_batch] == x],\n",
    "            \"Senescent_Cell_Label\", rankby_abs=True,\n",
    "            layer=\"log1p\", reference=\"Non-Senescent\", copy=True)\n",
    "    res_rank_genes[x][\"Overall\"] = sc.get.rank_genes_groups_df(\n",
    "            tmp, f\"Senescent\", pval_cutoff=p_threshold)\n",
    "    res_rank_genes[x] = pd.concat(res_rank_genes[x], names=[col_celltype])\n",
    "res_rank_genes = pd.concat(res_rank_genes, names=[col_batch])\n",
    "res_rank_genes = res_rank_genes.assign(\n",
    "    logfoldchanges_abs=res_rank_genes.logfoldchanges.abs())\n",
    "res_rank_genes_top = res_rank_genes.reset_index().groupby([\n",
    "    col_batch, col_celltype]).apply(\n",
    "        lambda x: x.sort_values(\"logfoldchanges_abs\").iloc[:20],\n",
    "        include_groups=False).reset_index(2, drop=True).set_index(\n",
    "            \"names\", append=True).rename_axis([\n",
    "                col_batch, col_celltype, \"variable\"])\n",
    "res_rank_genes_top"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "722ec3dc",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12d190c8",
   "metadata": {},
   "source": [
    "## Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adf21967",
   "metadata": {},
   "outputs": [],
   "source": [
    "scores = self.rna.obs[[col_batch] + [\n",
    "    f\"score_{x}\" for x in sen_metrics]].set_index(\n",
    "        col_batch, append=True).reset_index()\n",
    "sns.pairplot(scores, diag_kind=\"kde\", diag_kws=dict(\n",
    "    cut=0, fill=True, common_norm=True),\n",
    "             hue=col_batch, palette=palette[col_batch])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6d9261",
   "metadata": {},
   "source": [
    "## Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48080aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# UMAP with Senescence Scores\n",
    "bnds = self.rna.obs[\"senscore\"].describe(percentiles=[0.75, 0.99])\n",
    "_ = self.plot(kind=\"umap\", color=[\"senscore\", col_celltype], umap=dict(\n",
    "    palette=None, color_map=\"Reds\", vmin=bnds[\"75%\"], vmax=bnds[\"99%\"]))\n",
    "\n",
    "# Violin Plot of Senescence Scores\n",
    "_ = self.plot(kind=\"violin\", genes=[\"senscore\"], common_norm=True,\n",
    "              col_celltype=col_celltype, rotation=90)\n",
    "_ = self.plot(kind=\"violin\", genes=[\"senscore\"], common_norm=True,\n",
    "              by_group=col_celltype, col_wrap=1,\n",
    "              col_celltype=col_batch, rotation=90)\n",
    "\n",
    "# KDEs\n",
    "fig_sen_kde = {}\n",
    "for x in [col_age, col_condition, col_batch]:\n",
    "    fig_sen_kde[x] = sns.displot(self.rna.obs, hue=x, x=\"senscore\",\n",
    "                                 palette=palette[x], kind=\"kde\",\n",
    "                                 cut=0, common_norm=True, fill=True)\n",
    "\n",
    "# Overall (All Cell Types)\n",
    "fig_sen_overall = sns.catplot(self.rna.obs, x=col_condition,\n",
    "                              hue=col_age, y=\"senscore\",\n",
    "                              palette=palette[col_age],\n",
    "                              kind=\"violin\")\n",
    "fig_sen_overall_2 = sns.catplot(self.rna.obs, x=col_age, hue=col_condition,\n",
    "                                palette=palette[col_condition],\n",
    "                                y=\"senscore\", kind=\"violin\")\n",
    "\n",
    "# By Cell Type\n",
    "fig_sen_rc = sns.catplot(self.rna.obs, col=col_celltype, y=\"senscore\",\n",
    "                         palette=palette[col_condition],\n",
    "                         hue=col_condition, x=col_age, kind=\"violin\",\n",
    "                         height=10, aspect=2, col_wrap=3)\n",
    "fig_sen = sns.catplot(self.rna.obs, x=col_celltype, y=\"senscore\",\n",
    "                      palette=palette[col_batch],\n",
    "                      hue=col_batch, kind=\"box\", height=10, aspect=2)\n",
    "for a in fig_sen.axes.flatten():\n",
    "    a.tick_params(axis=\"x\", labelrotation=90)\n",
    "fig_sen_2 = sns.catplot(self.rna.obs, x=col_celltype, y=\"senscore\",\n",
    "                        col=col_age, hue=col_condition,\n",
    "                        palette=palette[col_condition],\n",
    "                        kind=\"box\", height=10, aspect=2)\n",
    "for a in fig_sen_2.axes.flatten():\n",
    "    a.tick_params(axis=\"x\", labelrotation=90)\n",
    "fig_sen_3 = sns.catplot(self.rna.obs, x=col_celltype, y=\"senscore\",\n",
    "                        col=col_condition, hue=col_age,\n",
    "                        palette=palette[col_age],\n",
    "                        kind=\"box\", height=10, aspect=2)\n",
    "for a in fig_sen_3.axes.flatten():\n",
    "    a.tick_params(axis=\"x\", labelrotation=90)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9a5f779",
   "metadata": {},
   "source": [
    "## Burden \n",
    "\n",
    "Group by sample and cell type and get percentage of senescent cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7f8d189",
   "metadata": {},
   "outputs": [],
   "source": [
    "grps = [col_sample, col_celltype, col_age, col_condition, col_batch]\n",
    "dff = self.rna.obs[grps + [\"Senescent_Cell\"]].groupby(grps).apply(\n",
    "    lambda x: x[\"Senescent_Cell\"].mean(), include_groups=False\n",
    "    ).to_frame(\"Senescent_Cell_Burden\") * 100\n",
    "# kws_plot = dict(kind=\"violin\", split=False, , common_norm=True)\n",
    "kws_plot = dict(kind=\"box\", errorbar=(\"ci\", 95))\n",
    "\n",
    "sns.catplot(y=\"Senescent_Cell_Burden\", x=col_condition, **kws_plot,\n",
    "            palette=palette[col_condition],\n",
    "            hue=col_condition, data=dff, height=3, aspect=1.5)\n",
    "sns.catplot(y=\"Senescent_Cell_Burden\", x=col_condition, **kws_plot,\n",
    "            col=col_celltype, col_wrap=4,\n",
    "            palette=palette[col_age],\n",
    "            hue=col_age, data=dff, height=3, aspect=1.5)\n",
    "sns.catplot(y=\"Senescent_Cell_Burden\", hue=col_condition, **kws_plot,\n",
    "            col=col_celltype, col_wrap=4,\n",
    "            palette=palette[col_condition],\n",
    "            x=col_age, data=dff, height=3, aspect=1.5)\n",
    "sns.catplot(y=\"Senescent_Cell_Burden\", x=col_celltype, **kws_plot,\n",
    "            palette=palette[col_batch],\n",
    "            hue=col_batch, data=dff, height=8, aspect=1.5)\n",
    "for x in [col_condition, col_age, [col_condition, col_age]]:\n",
    "    print(\"\\n\\n\", round(dff[\"Senescent_Cell_Burden\"].groupby(x).describe()[[\n",
    "        \"min\", \"25%\", \"50%\", \"75%\", \"max\"]], 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3089e717",
   "metadata": {},
   "source": [
    "# Analyze"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4ed1903",
   "metadata": {},
   "source": [
    "## Regressions & ANOVAs & MEMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a845670",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_dff = polars.DataFrame(dff.reset_index())  # pymer-compatible aggregated df\n",
    "r_rna = polars.DataFrame(self.rna.obs.rename({\n",
    "    col_celltype: \"CT_\"}, axis=1).copy())  # pymer-compatible cell-level df\n",
    "lvls = [\"key_control\", \"key_treatment\"]\n",
    "factors = {col_condition: [keys[col_condition][i] for i in lvls],\n",
    "           col_age: [keys[col_age][i] for i in lvls]}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a5227e0",
   "metadata": {},
   "source": [
    "### Scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f46a8af",
   "metadata": {},
   "source": [
    "##### Cell Type Random Effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fae8c3ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Age + Condition\n",
    "# model_ca_mlm = lmer(f\"senscore ~ {col_condition} + {col_age} +\"\n",
    "#                     f\"(1 | {col_sample})\", data=r_rna)  # set up model\n",
    "# model_ca_mlm.set_factors(factors)\n",
    "# model_ca_mlm.set_transforms({\"senscore\": \"zscore\"})\n",
    "# model_ca_mlm.anova(summary=True).show()\n",
    "\n",
    "# # Age * Condition\n",
    "# model_ca_mlm_i = lmer(\n",
    "#     f\"senscore ~ {col_condition} * {col_age} + \"\n",
    "#     f\"(1 | {col_sample})\", data=r_rna)  # set up model\n",
    "# model_ca_mlm_i.set_factors(factors)\n",
    "# model_ca_mlm_i.set_transforms({\"senscore\": \"zscore\"})\n",
    "# model_ca_mlm_i.anova(summary=True).show()\n",
    "\n",
    "# # Age * Condition (Condition as Random Effect)\n",
    "# model_ca_mlm_ii = lmer(\n",
    "#     f\"senscore ~ {col_condition} * {col_age} + \"\n",
    "#     f\"(1 | {col_sample}) + (1 + {col_condition} | CT_)\",\n",
    "#     data=r_rna)  # set up model\n",
    "# model_ca_mlm_ii.set_factors(factors)\n",
    "# model_ca_mlm_ii.set_transforms({\"senscore\": \"zscore\"})\n",
    "# model_ca_mlm_ii.anova(summary=True).show()\n",
    "\n",
    "# Age * Condition (Age and Condition as Random Effect)\n",
    "model_ca_mlm_iii = lmer(\n",
    "    f\"senscore ~ {col_condition} * {col_age} + (1 | {col_sample})\"\n",
    "    f\" + (1 + {col_condition} + {col_age} | CT_)\",\n",
    "    data=r_rna)  # set up model\n",
    "model_ca_mlm_iii.set_factors(factors)\n",
    "model_ca_mlm_iii.set_transforms({\"senscore\": \"zscore\"})\n",
    "model_ca_mlm_iii.anova(summary=True).show()\n",
    "\n",
    "# Model Comparison\n",
    "# compare(model_ca_mlm, model_ca_mlm_i, model_ca_mlm_i,\n",
    "#         model_ca_mlm_ii, model_ca_mlm_iii).show()\n",
    "\n",
    "# GLM Version of Results\n",
    "# model_ca_mlm.summary().show()\n",
    "# model_ca_mlm_i.summary().show()\n",
    "# model_ca_mlm_ii.summary().show()\n",
    "model_ca_mlm_iii.summary().show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "815c3d88",
   "metadata": {},
   "source": [
    "#### Cell Type as Moderator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e39a607",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Age + Condition * Cell Type\n",
    "model_cac_mlm_i2 = lmer(\n",
    "    f\"senscore ~  Aged + Spaceflight * CT_ + \"\n",
    "    f\"(1 | {col_sample})\", data=r_rna)  # set up model\n",
    "model_cac_mlm_i2.set_transforms({\"senscore\": \"zscore\"})\n",
    "model_cac_mlm_i2.anova(summary=True).show()\n",
    "\n",
    "# Age * Condition * Cell Type\n",
    "model_cac_mlm_i2b = lmer(\n",
    "    f\"senscore ~  Aged * Spaceflight + Spaceflight * CT_ + \"\n",
    "    f\"(1 | {col_sample})\", data=r_rna)  # set up model\n",
    "model_cac_mlm_i2b.set_transforms({\"senscore\": \"zscore\"})\n",
    "model_cac_mlm_i2b.anova(summary=True).show()\n",
    "\n",
    "# Age * Condition * Cell Type\n",
    "model_cac_mlm_i3 = lmer(\n",
    "    f\"senscore ~  Aged * Spaceflight * CT_ + \"\n",
    "    f\"(1 | {col_sample})\", data=r_rna)  # set up model\n",
    "model_cac_mlm_i3.set_transforms({\"senscore\": \"zscore\"})\n",
    "model_cac_mlm_i3.anova(summary=True).show()\n",
    "\n",
    "# GLM Version of Results\n",
    "model_cac_mlm_i2.summary().show()\n",
    "model_cac_mlm_i2b.summary().show()\n",
    "model_cac_mlm_i3.summary().show()\n",
    "\n",
    "# Model Comparison\n",
    "compare(model_cac_mlm_i2, model_cac_mlm_i2b, model_cac_mlm_i3).show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8009a21",
   "metadata": {},
   "source": [
    "### Binary Snc/Not\n",
    "\n",
    "Multilevel Logistic Regression"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "478b409c",
   "metadata": {},
   "source": [
    "#### Cell Type Random Effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b75eb47e",
   "metadata": {},
   "outputs": [],
   "source": [
    "glmmod = {}\n",
    "# for x in [\"1\", f\"1 + {col_condition}\", f\"1 + {col_condition} + {col_age}\"]:\n",
    "#     fff = str(f\"Senescent_Cell ~ {col_age} * {col_condition}\"\n",
    "#               f\" + (1 | {col_sample}) + ({x} | CT_)\")\n",
    "for x in [\"1\", f\"1 + Spaceflight\", f\"1 + Spaceflight + Aged\"]:\n",
    "    sep = \"|\" if x == \"1\" else \"||\"\n",
    "    fff = str(f\"Senescent_Cell ~ Aged * Spaceflight\"\n",
    "              f\" + (1 | {col_sample}) + ({x} {sep} CT_)\")\n",
    "    glmmod[x] = glmer(fff, data=r_rna, family=\"binomial\")\n",
    "    # glmmod[x].set_factors(factors)\n",
    "    glmmod[x].fit(exponentiate=True, ncpus=ncpus)  # estimates => odds\n",
    "    # print(glmmod[x].emmeans(col_condition))\n",
    "    # print(glmmod[x].emmeans(col_age))\n",
    "    glmmod[x].summary().show()\n",
    "    # print(glmmod[x].ranef[\"CT_\"])\n",
    "compare(*[glmmod[x] for x in glmmod])\n",
    "\n",
    "# Plot\n",
    "# grid = sns.FacetGrid(self.rna.obs, col=col_age, hue=col_condition)\n",
    "# grid.map(sns.countplot, \"Senescent_Cell\", order=[False, True])\n",
    "# grid.add_legend()\n",
    "# grid.set_xlabels(\"\")\n",
    "# grid.set_xticklabels([\"Not SnC\", \"Snc\"])\n",
    "# sns.despine()\n",
    "# fig = plt.figure(figsize=(10, 2))\n",
    "# g_s = fig.add_gridspec(1, 2, width_ratios=[1, 3])\n",
    "# left, right = fig.add_subplot(g_s[0]), fig.add_subplot(g_s[1])\n",
    "# left = sns.kdeplot(x=\"resid\", data=glmmod.data, ax=left)\n",
    "# left.set(xlabel=\"Residuals\", title=\"Error Distribution\")\n",
    "# right.stem(glmmod.data[\"cooksd\"], basefmt=\" \", label=\"Cook's Distance\")\n",
    "# right.set(xlabel=\"Row Number\", ylabel=\"Cook's Distance\",\n",
    "#           title=\"Influential Observations\",\n",
    "#           xticks=range(glmmod.data.height))\n",
    "# sns.despine()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a29efafa",
   "metadata": {},
   "source": [
    "#### Cell Type as Moderator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d728ddc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2-Way Interaction\n",
    "fff = str(f\"Senescent_Cell ~ Aged + Spaceflight * CT_ + \"\n",
    "          \"(1 | sample)\")\n",
    "glmmod_ct2 = glmer(fff, data=r_rna, family=\"binomial\")\n",
    "glmmod_ct2.fit(exponentiate=True, ncpus=ncpus)  # estimates => odds\n",
    "glmmod_ct2.summary().show()\n",
    "\n",
    "# 2 2-Way Interactions\n",
    "fff = str(f\"Senescent_Cell ~  Aged * Spaceflight + Spaceflight * CT_\"\n",
    "          f\" + (1 | {col_sample})\")\n",
    "glmmod_ct2b = glmer(fff, data=r_rna, family=\"binomial\")\n",
    "glmmod_ct2b.fit(exponentiate=True, ncpus=ncpus)  # estimates => odds\n",
    "glmmod_ct2b.anova(summary=True).show()\n",
    "glmmod_ct2b.summary().show()\n",
    "\n",
    "# 3-Way Interaction\n",
    "fff = str(f\"Senescent_Cell ~  Aged * Spaceflight * CT_\"\n",
    "          f\" + (1 | {col_sample})\")\n",
    "model_allix = glmer(fff, data=r_rna, family=\"binomial\")\n",
    "model_allix.fit(exponentiate=True, ncpus=ncpus)  # estimates => odds\n",
    "model_allix.anova(summary=True).show()\n",
    "model_allix.summary().show()\n",
    "res_allix_fixed = pd.DataFrame(model_allix.result_fit[[\n",
    "    \"term\", \"estimate\", \"p_value\"]], columns=[\n",
    "        \"Term\", \"OR\", \"P\"]).set_index(\"Term\")\n",
    "res_allix_fixed.loc[:, \"Log-Odds\"] = np.log(\n",
    "    res_allix_fixed.OR.astype(float))\n",
    "res_allix_fixed.loc[:, \"Significance\"] = res_allix_fixed.P.apply(\n",
    "    lambda x: \"***\" if x < 0.001 else \"**\" if x < 0.01 else \"*\" if (\n",
    "        x < 0.05) else \"\")\n",
    "print(res_allix_fixed)\n",
    "\n",
    "# Model Comparison\n",
    "compare(glmmod_ct2, glmmod_ct2b, model_allix)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8fbecf5",
   "metadata": {},
   "source": [
    "### Burden (Percent SnC)\n",
    "\n",
    "ANOVA: Percent SnCs across (a) Condition x Age and (b) Condition x Age x Cell Type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3f36a24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Age + Condition\n",
    "model_1 = lm(\n",
    "    f\"Senescent_Cell_Burden ~ {col_condition} + {col_age}\", data=r_dff)\n",
    "model_1.set_factors(factors)\n",
    "model_1.anova(summary=True).show()\n",
    "model_1.summary().show()\n",
    "\n",
    "# Age + Condition +  Cell Type\n",
    "model_2 = lm(\n",
    "    f\"Senescent_Cell_Burden ~ {col_condition} + {col_age} + {col_celltype}\",\n",
    "    data=r_dff)\n",
    "model_2.set_factors(factors)\n",
    "model_2.anova(summary=True).show()\n",
    "model_2.summary().show()\n",
    "\n",
    "# Age + Condition * Cell Type\n",
    "model_3 = lm(\n",
    "    f\"Senescent_Cell_Burden ~ {col_age} + {col_condition} * {col_celltype}\",\n",
    "    data=r_dff)\n",
    "model_3.set_factors(factors)\n",
    "model_3.anova(summary=True).show()\n",
    "\n",
    "# # Age * Condition * Cell Type\n",
    "# model_4 = lm(\n",
    "#     f\"Senescent_Cell_Burden ~ {col_age} * {col_condition} * {col_celltype}\",\n",
    "#     data=r_dff)\n",
    "# model_4.set_factors(factors)\n",
    "# model_4.anova(summary=True).show()\n",
    "# model_4.summary().show()\n",
    "\n",
    "# Compare\n",
    "# compare(model_1, model_2, model_3, model_4).show()\n",
    "for x in [model_1, model_2, model_3]:\n",
    "    x.summary().show()\n",
    "compare(model_1, model_2, model_3).show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03a93350",
   "metadata": {},
   "source": [
    "### Just Age as Predictor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bfbbe3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model_cac_mlm_i3_a = lmer(\n",
    "#     f\"senscore ~  {col_age} * {col_celltype} + (1 | {col_sample})\",\n",
    "#     data=r_rna)  # set up model\n",
    "# model_cac_mlm_i3_a.set_factors(factors)\n",
    "# model_cac_mlm_i3_a.set_transforms({\"senscore\": \"zscore\"})\n",
    "# model_cac_mlm_i3_a.fit()\n",
    "# model_cac_mlm_i3_a.summary().show()\n",
    "\n",
    "# fff = str(f\"Senescent_Cell ~ Aged * {col_celltype} + \"\n",
    "#           \"(1 | sample)\")\n",
    "# glmmod_ct2a = glmer(fff, data=r_rna, family=\"binomial\")\n",
    "# glmmod_ct2a.fit(exponentiate=True, ncpus=ncpus)  # estimates => odds\n",
    "# glmmod_ct2a.summary().show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "885efb6c",
   "metadata": {},
   "source": [
    "### Lower SnC Threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0465ca2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Column for Lower Threshold Classification\n",
    "chp = f\"Senescent_Cell_{use_metric}_{p_h}\"  # column name\n",
    "\n",
    "# Cell Type Random Effects\n",
    "fff = str(f\"{chp} ~ Aged * Spaceflight\"\n",
    "          f\" + (1 | {col_sample}) + (1 + Spaceflight + Aged || CT_)\")\n",
    "glmmod_lenient_thresh = glmer(fff, data=r_rna, family=\"binomial\")\n",
    "glmmod_lenient_thresh.fit(exponentiate=True, ncpus=ncpus)  # => odds\n",
    "glmmod_lenient_thresh.summary().show()\n",
    "\n",
    "# Cell Type as Moderator\n",
    "fff = str(f\"{chp} ~ Aged * Spaceflight * CT_ + (1 | {col_sample})\")\n",
    "glmmod_lenient_thresh_ct = glmer(fff, data=r_rna, family=\"binomial\")\n",
    "glmmod_lenient_thresh_ct.fit(exponentiate=True, ncpus=ncpus)  # => odds\n",
    "glmmod_lenient_thresh_ct.summary().show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0692eeb",
   "metadata": {},
   "source": [
    "## Perturbation Distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba681854",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"\\n\\n{'=' * 80}\\nOverall Distance\\n{'=' * 80}\\n\\n\")\n",
    "df_distance, figs_distance = scflow.ax.analyze_perturbation_distance(\n",
    "    self.rna, col_condition=[col_age, col_condition, col_batch])\n",
    "\n",
    "print(f\"\\n\\n{'=' * 80}\\nJust Senescence Genes\\n{'=' * 80}\\n\\n\")\n",
    "df_distance_snc, figs_distance_snc = scflow.ax.analyze_perturbation_distance(\n",
    "    self.rna[:, genes], col_condition=[\n",
    "        col_age, col_condition, col_batch])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c9997e9",
   "metadata": {},
   "source": [
    "## Composition Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb50e62f",
   "metadata": {},
   "source": [
    "### Cell Type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74069908",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(f\"\\n\\n{'=' * 80}\\nOverall Sample\\n{'=' * 80}\\n\\n\")\n",
    "# out_ctc = scflow.ax.analyze_composition(\n",
    "#     self.rna, col_celltype, col_batch, col_sample=col_sample,\n",
    "#     palette=palette[col_batch],\n",
    "#     formula=None, key_modality=\"coda\", reference_cell_type=\"automatic\",\n",
    "#     absence_threshold=0.1, est_fdr=0.1, level_order=[keys[\n",
    "#         col_batch][\"key_control\"]] + keys[col_batch][\"key_treatment\"])\n",
    "# plt.show()\n",
    "\n",
    "# print(f\"\\n\\n{'=' * 80}\\nAged Subset\\n{'=' * 80}\\n\\n\")\n",
    "# out_ctc_sf = scflow.ax.analyze_composition(\n",
    "#     self.rna[self.rna.obs[col_age] != keys[col_age][\"key_control\"]],\n",
    "#     col_celltype, col_condition, col_sample=col_sample,\n",
    "#     formula=None, key_modality=\"coda\", reference_cell_type=\"automatic\",\n",
    "#     absence_threshold=0.1, est_fdr=0.1, level_order=[keys[col_condition][\n",
    "#         \"key_control\"]] + [keys[col_condition][\"key_treatment\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa872982",
   "metadata": {},
   "source": [
    "### Senescent Cells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25be8590",
   "metadata": {},
   "source": [
    "#### Tree-Aggregated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afb3fdfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup\n",
    "# col_celltype_hierarchy = [\n",
    "#     f\"{col_celltype}_hierarchical\", \"SnC_hierarchy\"]\n",
    "col_celltype_hierarchy = [\n",
    "    \"Senescent_Cell\", f\"{col_celltype}_hierarchical\", \"SnC_hierarchy\",\n",
    "    \"Senescent_Cell_Label_by_Type\"]\n",
    "# col_celltype_hierarchy = [\n",
    "#     \"Senescent_Cell_Label\", f\"SnC_hierarchy\",\n",
    "#     \"Senescent_Cell_Label_by_Type\"]\n",
    "ref_celltype = self.rna.obs[col_celltype_hierarchy[-1]].value_counts(\n",
    "    ).index.values[0]\n",
    "form_c = \" + \".join([col_condition, col_age]) + \" + \" + \" * \".join([\n",
    "    col_condition, col_age])\n",
    "\n",
    "# By SF vs. GC (No Age)\n",
    "_ = scflow.ax.analyze_composition_tree(\n",
    "    self.rna, col_celltype_hierarchy[-1], col_condition,\n",
    "    col_sample=col_sample, formula=None, seed=0, est_fdr=0.2,\n",
    "    reference_cell_type=ref_celltype, num_warmup=2000,\n",
    "    col_celltype_hierarchy=col_celltype_hierarchy)\n",
    "\n",
    "# By SF & Age\n",
    "_ = scflow.ax.analyze_composition_tree(\n",
    "    self.rna, col_celltype_hierarchy[-1], [col_condition, col_age],\n",
    "    formula=form_c, col_sample=col_sample, seed=0,\n",
    "    reference_cell_type=ref_celltype, est_fdr=0.2,\n",
    "    col_celltype_hierarchy=col_celltype_hierarchy)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cce35182",
   "metadata": {},
   "source": [
    "#### `scCoda`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36577580",
   "metadata": {},
   "source": [
    "##### Just by Condition (By Cell Type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "843faa54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for x in self.rna.obs[col_celltype].unique():\n",
    "#     r_s, num_samples, num_warmup = 0, 10000, 2000  # defaults\n",
    "#     if \"annotation_scanvi\" in col_celltype:  # avoid bad acceptance rates\n",
    "#         r_s = 16 if x  == \"OPC\" else 8675309 if x in [\n",
    "#             \"Microglial\"] else r_s  # different random seed\n",
    "#         if x == \"Endothelial\":\n",
    "#             r_s, num_samples, num_warmups = 7654321, 14000, 2000\n",
    "#     print(f\"Seed = {r_s}; {num_samples} samples; {num_samples} warmup\")\n",
    "#     out = scflow.ax.analyze_composition(\n",
    "#         self.rna[self.rna.obs[col_celltype] == x],\n",
    "#         \"Senescent_Cell_Label_by_Type\", col_condition,\n",
    "#         col_sample=col_sample, formula=None, key_modality=\"coda\",\n",
    "#         full_hmc=False, palette=palette[col_condition], seed=r_s,\n",
    "#         num_samples=10000, num_warmup=2000,\n",
    "#         reference_cell_type=x, absence_threshold=0.1, est_fdr=0.2,\n",
    "#         level_order=[keys[col_condition][\"key_control\"], keys[\n",
    "#             col_condition][\"key_treatment\"]])\n",
    "#     f_sccoda_out = f\"outputs/sccoda_results_{col_celltype}_{x}_sfgc.pkl\"\n",
    "#     if overwrite is True or os.path.exists(f_sccoda_out) is False:\n",
    "#         with open(f_sccoda_out, \"wb\") as f:\n",
    "#             pickle.dump(out[0], f)\n",
    "#         # out[1].write_h5mu(\n",
    "#         #     f\"outputs/sccoda_data_{col_celltype}_{x}_sfgc.h5mu\")\n",
    "#     plt.show()\n",
    "#     del out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bec511c",
   "metadata": {},
   "source": [
    "##### By Condition & Age"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f378185c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d90dd71",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"\\n\\n{'=' * 80}\\nBy Cell Type (by Group)\\n{'=' * 80}\\n\\n\")\n",
    "for x in self.rna.obs[col_celltype].unique():\n",
    "    r_s = 1618\n",
    "    if \"annotation_scanvi\" in col_celltype:\n",
    "        r_s = 32145 if x in [\"Neuron\"] else 16 if x in [\n",
    "            \"Microglia\", \"OPC\"] else r_s\n",
    "    out = scflow.ax.analyze_composition(\n",
    "        self.rna[self.rna.obs[col_celltype] == x],\n",
    "        \"Senescent_Cell_Label_by_Type\", col_batch, col_sample=col_sample,\n",
    "        formula=None, key_modality=\"coda\", full_hmc=False,\n",
    "        palette=palette[col_batch], seed=r_s,\n",
    "        num_samples=10000, num_warmup=2000,\n",
    "        reference_cell_type=x, absence_threshold=0.1, est_fdr=0.2,\n",
    "        level_order=[keys[col_batch][\"key_control\"]] + keys[\n",
    "            col_batch][\"key_treatment\"])\n",
    "    f_sccoda_out = f\"outputs/sccoda_results_{col_celltype}_{x}.pkl\"\n",
    "    if overwrite is True or os.path.exists(f_sccoda_out) is False:\n",
    "        with open(f_sccoda_out, \"wb\") as f:\n",
    "            pickle.dump(out[0], f)\n",
    "        # out[1].write_h5mu(f\"outputs/sccoda_data_{col_celltype}_{x}.h5mu\")\n",
    "    plt.show()\n",
    "    del out\n",
    "\n",
    "# Reload\n",
    "# out_ctc_ct = {}\n",
    "# for x in self.rna.obs[col_celltype].unique():\n",
    "#     with open(f\"outputs/sccoda_results_{col_celltype}_{x}.pkl\", \"rb\") as f:\n",
    "#         tmp = pickle.load(f)\n",
    "#     out_ctc_ct[x] = [tmp, mudata.read(\n",
    "#         f\"outputs/sccoda_data_{col_celltype}_{x}.h5mu\")]\n",
    "#     del tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1535a760",
   "metadata": {},
   "source": [
    "##### Lower SnC Threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77414f69",
   "metadata": {},
   "outputs": [],
   "source": [
    "chp = f\"Senescent_Cell_Label_by_Type_{use_metric}_{p_h}\"  # column name\n",
    "print(\"Cross-Tabs of Cell Counts in SnC Classifications \"\n",
    "      f\"with Top {percentile}% Threshold versus Top {p_h}%\")\n",
    "print(self.rna.obs[[\"Senescent_Cell_Label_by_Type\", chp]].value_counts(\n",
    "    ).sort_index(), \"\\n\\n\")  # cross-tabs of higher vs. lower SnC threshold\n",
    "# print(self.rna.obs[[\"Senescent_Cell_Label_by_Type\", chp, col_batch]].groupby(\n",
    "#     col_batch).value_counts().unstack(0).replace(0, np.nan).dropna(\n",
    "#         how=\"all\", axis=0).astype(int).sort_index())  # cross-tabs by group\n",
    "for x in self.rna.obs[col_celltype].unique():\n",
    "    out = scflow.ax.analyze_composition(\n",
    "        self.rna[self.rna.obs[col_celltype] == x],\n",
    "        chp, col_condition, col_sample=col_sample,\n",
    "        formula=None, key_modality=\"coda\", full_hmc=False,\n",
    "        palette=palette[col_condition], seed=1618,\n",
    "        num_samples=10000, num_warmup=2000,\n",
    "        reference_cell_type=x, absence_threshold=0.1, est_fdr=0.2,\n",
    "        level_order=[keys[col_condition][\"key_control\"], keys[\n",
    "            col_condition][\"key_treatment\"]])\n",
    "    f_sccoda = f\"outputs/sccoda_results_{col_celltype}_{x}_sfgc_{p_h}pct.pkl\"\n",
    "    if overwrite is True or os.path.exists(f_sccoda) is False:\n",
    "        with open(f_sccoda, \"wb\") as f:\n",
    "            pickle.dump(out[0], f)\n",
    "        out[1].write_h5mu(\n",
    "            f\"outputs/sccoda_data_{col_celltype}_{x}_sfgc_{p_h}pct.h5mu\")\n",
    "    plt.show()\n",
    "    del out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3255006",
   "metadata": {},
   "source": [
    "## Differential Gene Expression\n",
    "\n",
    "Mainly using `edgeR`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a698f9bb",
   "metadata": {},
   "source": [
    "### Pseudo-Bulk & Examine Covariates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67479d29",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata = scflow.tl.create_pseudobulk(\n",
    "    self.rna, [i for i in [col_sample, col_condition, col_age] if i],\n",
    "    \"Senescent_Cell_Label_by_Type\", layer=\"counts\", mode=\"sum\")\n",
    "pdata.layers[\"counts\"] = pdata.X.copy()\n",
    "sc.pp.normalize_total(pdata, target_sum=1e6)\n",
    "sc.pp.log1p(pdata)\n",
    "sc.pp.pca(pdata)\n",
    "sc.pl.pca(pdata, color=pdata.obs, ncols=1, size=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "426c47a5",
   "metadata": {},
   "source": [
    "### Overall DEGs by Groups"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caaabdc0",
   "metadata": {},
   "source": [
    "#### Age\n",
    "\n",
    "**In older spaceflight mice compared to younger spaceflight mice**\n",
    "* 9630013a20Rik (CNS myelination and neural tissue regeneration) is downregulated.\n",
    "* mt-Td is upregulated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f87fd6a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "# Overall\n",
    "print(f\"\\n\\n{'=' * 80}\\nOverall Sample\\n{'=' * 80}\\n\\n\")\n",
    "out_edgr_age = scflow.ax.run_deg_edgr(\n",
    "    self.rna, col_age, col_covariate=None, formula=None,\n",
    "    **keys[col_age], log2fc_thresh=0, n_top_vars=25,\n",
    "    col_celltype=col_celltype, col_sample=col_sample,\n",
    "    fig_title=\"Age DEGs: Overall Sample\\n\\n\")\n",
    "\n",
    "# Ground Control Subset\n",
    "print(f\"\\n\\n{'=' * 80}\\nGround Control Subset\\n{'=' * 80}\\n\\n\")\n",
    "out_edgr_age_gc = scflow.ax.run_deg_edgr(\n",
    "    self.rna[self.rna.obs[col_condition] == keys[\n",
    "        col_condition][\"key_control\"]],\n",
    "    col_age, col_covariate=None, formula=None,\n",
    "    **keys[col_age], log2fc_thresh=0, n_top_vars=25,\n",
    "    col_celltype=col_celltype, col_sample=col_sample,\n",
    "    fig_title=\"Age DEGs: Just Ground Control\\n\\n\")\n",
    "\n",
    "# Space Flight Subset\n",
    "print(f\"\\n\\n{'=' * 80}\\nSpace Flight Subset\\n{'=' * 80}\\n\\n\")\n",
    "out_edgr_age_sf = scflow.ax.run_deg_edgr(\n",
    "    self.rna[self.rna.obs[col_condition] == keys[\n",
    "        col_condition][\"key_treatment\"]],\n",
    "    col_age, col_covariate=None, formula=None, **keys[col_age],\n",
    "    log2fc_thresh=0, n_top_vars=25, col_celltype=col_celltype,\n",
    "    col_sample=col_sample, fig_title=\"Age DEGs: Just Space Flight\\n\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8835196",
   "metadata": {},
   "source": [
    "#### Condition\n",
    "\n",
    "```\n",
    "out_edgr_con_o[0][(out_edgr_con_o[0].adj_p_value < 0.01) & (\n",
    "    out_edgr_con_o[0].abs_log_fc > 0.5)].sort_values(\n",
    "        \"abs_log_fc\", ascending=False)\n",
    "```\n",
    "\n",
    "Among the older mice, *Mir6236* was differentially highly expressed in aged spaceflight mice compared to ground control mice (LFC=1.95, adjusted p=2.61e-13 in the overall sample; LFC=2.11, adjusted p=0.000015 in the aged subset). Suppression of this gene has been linked to neuronal morphogenesis (https://pubs.acs.org/doi/abs/10.1021/acsabm.0c01389) and therefore may represent a candidate intervention target. \n",
    "\n",
    "*Il31ra*, which has been linked to neuroinflammation (https://advanced.onlinelibrary.wiley.com/doi/pdfdirect/10.1002/advs.202409086), was also more highly expressed in spaceflight mice compared to ground control mice (LFC=1.52, adjusted p<0.001 in the overall sample; LFC=1.53, adjusted p<0.001 in the aged subset).\n",
    "\n",
    "*UBA52*, also more highly expressed in spaceflight mice (LFC=1.43, adjusted p<0.001 in the overall sample; LFC=1.45, adjusted p<0.001 in the aged subset), may serve a neuroprotective role: It has been shown underexpressed in PD (https://www.mdpi.com/2073-4409/11/23/3770) and plays a role in neuroautophagy (https://www.tandfonline.com/doi/pdf/10.1080/15548627.2024.2395727). Likewise, *Pomp*, a gene associated with oxidative stress resistance and protein clearance [CITATION] was also more highly expressed in spaceflight mice. In contrast to past observations of downregulation of *Uqcr11* in Alheimer's (https://www.benthamdirect.com/content/journals/car/10.2174/1567205014666170505095921), that gene was differentially highly expressed in spaceflight mice (LFC=1.23 aged subset; LFC=1.44 overall; p<0.001); however, *Cmss1*, which has been associated with Alzheimer's in APOE4 mice (https://pmc.ncbi.nlm.nih.gov/articles/PMC8064208/), was also more highly expressed in spaceflight mice.\n",
    "\n",
    "Genes involved in oxidative phosphorylation and the electron transport chain, including *Atp5e*, *Ndufa12*, and *Uqcr11* were also more highly expressed in spaceflight mice. *Uqcr11* has shown an anti-apoptotic and oxidative stress alleviation role in mice (https://www.mdpi.com/2073-4425/16/5/526).\n",
    "\n",
    "*Tpt1* and *Rn7sk* were also upregulated in spaceflight mice. *Tpt1*, which is associated with cancer progression, anti-apoptosis, and cell stress response, has been shown to have a potentially bi-directional relationship with cell cycle control gene and canonical senescence marker *TP53* (https://www.tandfonline.com/doi/pdf/10.4161/cc.25404). *Rn7sk* is associated with cellular senescence (https://onlinelibrary.wiley.com/doi/abs/10.1002/jcp.28119) and Alzheimer's (https://www.nature.com/articles/s41598-024-82490-9.pdf).\n",
    "\n",
    "---\n",
    "\n",
    "*Pvalb* overall sample (LFC=1.61, p=1.25e-11)\n",
    "\n",
    "*Trim17*\n",
    "\n",
    "*Cck*\n",
    "\n",
    "*Pam16*\n",
    "\n",
    "---\n",
    "\n",
    "*Mitochondrial/electron transport chain genes whose upregulation has been linked to Parkinson's disease (https://pmc.ncbi.nlm.nih.gov/articles/PMC10259399/pdf/AMS-19-3-131629.pdf) and/or adaptive responses to neurodegeneration in Alheimer's mouse models (https://journals.sagepub.com/doi/abs/10.1177/13872877251314847), including *mt-Atp6*, *mt-Co1*, *mt-Cytb* *mt-Nd2*, and *mt-Nd5*, *mt-Nd6...pregulation of *mt-Rnr1*, which also showed higher expression in spaceflight aged mice, has been seen in response to ionizing radiation, possibly reflecting increased mitochonrial biogenesis as an adaptive stress response (https://pmc.ncbi.nlm.nih.gov/articles/PMC5555881)** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1194ca7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Overall\n",
    "print(f\"\\n\\n{'=' * 80}\\nOverall Sample\\n{'=' * 80}\\n\\n\")\n",
    "out_edgr_con = scflow.ax.run_deg_edgr(\n",
    "    self.rna, col_condition, col_covariate=None, formula=None,\n",
    "    **keys[col_condition], log2fc_thresh=0, n_top_vars=25,\n",
    "    col_celltype=col_celltype, col_sample=col_sample,\n",
    "    fig_title=f\"{col_condition} DEGs (Overall Sample)\\n\\n\")\n",
    "out_edgr_con_df = out_edgr_con[0][out_edgr_con[\n",
    "    0].adj_p_value < 0.001].sort_values(\"abs_log_fc\", ascending=False)\n",
    "goe = out_edgr_con_df.variable[:10].to_list()\n",
    "print(out_edgr_con_df.iloc[:min(100, out_edgr_con_df.shape[0])], \"\\n\\n\", goe)\n",
    "plt.show()\n",
    "\n",
    "# Just Old\n",
    "print(f\"\\n\\n{'=' * 80}\\nAged Subset\\n{'=' * 80}\\n\\n\")\n",
    "out_edgr_con_o = scflow.ax.run_deg_edgr(\n",
    "    self.rna[self.rna.obs[col_age] == keys[col_age][\"key_treatment\"]],\n",
    "    col_condition, col_covariate=None, formula=None,\n",
    "    **keys[col_condition], log2fc_thresh=0, n_top_vars=25,\n",
    "    col_celltype=col_celltype, col_sample=col_sample,\n",
    "    fig_title=f\"{col_condition} DEGs (Just Aged)\\n\\n\")\n",
    "plt.show()\n",
    "out_edgr_con_o_df = out_edgr_con_o[0][out_edgr_con_o[\n",
    "    0].adj_p_value < 0.001].sort_values(\"abs_log_fc\", ascending=False)\n",
    "goes = out_edgr_con_o_df.variable[:10].to_list()\n",
    "print(out_edgr_con_o_df, \"\\n\\n\", goes)\n",
    "_ = self.plot(\n",
    "    kind=[\"violin\", \"matrix\"], col_celltype=col_batch, layer=\"log1p\",\n",
    "    violin=dict(col_wrap=5, hspace=0.25, rotation=45, figsize=(20, 20)),\n",
    "    matrix=dict(standard_scale=\"var\"), genes=goes)\n",
    "out_edgr_old_sf_v_gc = out_edgr_con_o_df[0][(\n",
    "    out_edgr_con_o_df[0].abs_log_fc > 1) & (\n",
    "        out_edgr_con_o_df[0].adj_p_value < 0.001)].sort_values(\n",
    "            \"adj_p_value\", ascending=True)\n",
    "out_edgr_old_sf_v_gc = out_edgr_old_sf_v_gc.groupby(\"contrast\").apply(\n",
    "    lambda x: x.sort_values(\"adj_p_value\", ascending=True).iloc[\n",
    "        :20]).set_index(\"variable\", append=True).reset_index(\n",
    "            1, drop=True).rename_axis([col_condition, \"variable\"])\n",
    "out_edgr_con_compare = out_edgr_con_o_df.set_index(\"variable\")[[\n",
    "    \"log_fc\", \"adj_p_value\"]].join(out_edgr_con_df.set_index(\"variable\")[[\n",
    "        \"log_fc\", \"adj_p_value\"]], how=\"inner\", lsuffix=\"_Aged_Subset\",\n",
    "                                   rsuffix=\"_Overall_Sample\")\n",
    "out_edgr_con_compare"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "181704aa",
   "metadata": {},
   "source": [
    "#### Age x Condition\n",
    "\n",
    "Spaceflight effect (main analysis), can visualize at different ages\n",
    "\n",
    "*Chchd2* (related to oxidative phosphorylation, cell migration, apoptosis inhibition, COX regulation, stress adaptation, Parkinson's, Lewy Body diseases, and cancer) was more highly expressed in spaceflight mice (and demonstrated a bigger young versus old difference in spaceflight mice).\n",
    "Spaceflight mice also showed higher expression of several ribosomal proteins (*Rps16*, *Rpl31*, *Rpl17*, *Rpl9*, *Rpl15*, *Rps15a*, *Rpl19*, *Rps7*, *Rps23*, *Rpl37*, *Rps3a1*, *Rpl29*, *Rps15*, *Rps14*).\n",
    "\n",
    "*Ndufa4* was also more highly expressed (implicated in cancer and mitochondrial dysfunction, part of the cytochrome c oxidase complex which drives oxidative phosphorylation, adaptation to hypoxia via downregulation of oxygen consumption and suppression of mitochondrial ROS production https://www.cell.com/cell-metabolism/fulltext/S1550-4131(11)00394-9?mobileUi=0) in spaceflight mice, as were other genes involved in ATP production/the electron transport chain and oxidative phosophorylation, including *Cox7a2*, *Uqcr11*, *Ndufa12*, *Ndufa4*, *Ndufc1*, *Atp5o-1*, *Atp5e*, and *Uqcrh* (which additionally has been implicated in tumor suppression https://aacrjournals.org/cancerres/article/80/16_Supplement/4795/643367).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f513f078",
   "metadata": {},
   "outputs": [],
   "source": [
    "# With Single Label for Age x Condition\n",
    "print(f\"\\n\\n{'=' * 80}\\nGroup (Single Factor)\\n{'=' * 80}\\n\\n\")\n",
    "out_edgr_groups = scflow.ax.run_deg_edgr(\n",
    "    self.rna, col_batch, col_covariate=None, formula=None,\n",
    "    **keys[col_batch], col_sample=col_sample,\n",
    "    log2fc_thresh=0, n_top_vars=25, col_celltype=col_celltype)\n",
    "print(list(out_edgr_groups[0].variable[:25]))\n",
    "plt.show()\n",
    "# _ = self.plot(kind=\"violin\", col_celltype=col_batch,\n",
    "#               genes=list(out_edgr_groups[0].variable[:25]), layer=\"scaled\",\n",
    "#               violin=dict(col_wrap=4, hspace=1, rotation=45))\n",
    "out_edgr_batches = out_edgr_groups[0][(out_edgr_groups[0].abs_log_fc > 1) & (\n",
    "    out_edgr_groups[0].adj_p_value < 0.001)].sort_values(\n",
    "        \"adj_p_value\", ascending=True)\n",
    "out_edgr_batches_top = out_edgr_batches.groupby(\"contrast\").apply(\n",
    "    lambda x: x.sort_values(\"adj_p_value\", ascending=True).iloc[\n",
    "        :20]).set_index(\"variable\", append=True).reset_index(\n",
    "            1, drop=True).rename_axis([col_condition, \"variable\"])\n",
    "\n",
    "# With Contrasts (Age & Condition = Separate Factors)\n",
    "print(f\"\\n\\n{'=' * 80}\\nAge * Condition (2-Factor Contrast)\\n{'=' * 80}\\n\\n\")\n",
    "out_edgr_contrasts = scflow.ax.run_deg_edgr(\n",
    "    self.rna, col_condition, col_covariate=col_age, formula=None,\n",
    "    **keys[col_condition], log2fc_thresh=0, n_top_vars=24,\n",
    "    col_celltype=col_celltype, col_sample=col_sample)\n",
    "_ = self.plot(\n",
    "    kind=[\"violin\", \"matrix\"], col_celltype=col_batch,\n",
    "    genes=out_edgr_contrasts[0][out_edgr_contrasts[\n",
    "        0].adj_p_value < 0.001].sort_values(\n",
    "            \"abs_log_fc\", ascending=False).variable[:10], layer=\"log1p\",\n",
    "    violin=dict(col_wrap=5, hspace=0.25, rotation=45, figsize=(20, 20)),\n",
    "    matrix=dict(standard_scale=\"var\"))\n",
    "\n",
    "# Condition x Age\n",
    "# Vulnerability of mice who were older at launch to spaceflight stress\n",
    "out_edgr_contrasts_age = scflow.ax.run_deg_edgr(\n",
    "    self.rna, col_age, col_covariate=col_condition, formula=None,\n",
    "    **keys[col_age], log2fc_thresh=0, n_top_vars=24,\n",
    "    col_celltype=col_celltype, col_sample=col_sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfd4f0a5",
   "metadata": {},
   "source": [
    "#### Senescence Gene Contrasts Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc393114",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genes\n",
    "sen_genes_in_contrasts = list(set(genes).intersection(\n",
    "    out_edgr_contrasts[0].variable.unique()))  # senescence genes in contrasts\n",
    "out_edgr_snc = out_edgr_contrasts[0].set_index(\n",
    "    \"variable\").loc[sen_genes_in_contrasts]  # contrast results for snc genes\n",
    "res_snc_contrast = out_edgr_snc[out_edgr_snc.adj_p_value < 0.01]  # p < 0.01\n",
    "sig_contrast_genes = list(res_snc_contrast.index.values)  # significant genes\n",
    "\n",
    "# Plots By Groups; Axis by Cell Types\n",
    "kws = dict(dot=dict(figsize=(10, 5)), heat=dict(figsize=(15, 15)))\n",
    "_ = self.plot(kind=[\"matrix\", \"dot\"], genes=sig_contrast_genes,\n",
    "              standard_scale=\"obs\", col_celltype=col_celltype,\n",
    "              by_group=col_batch, **kws)\n",
    "\n",
    "# Plots By Cell Types; Axis by Groups\n",
    "# kws = dict(dot=dict(figsize=(10, 5)), heat=dict(figsize=(15, 15)),\n",
    "#            violin=dict(figsize=(30, 20), col_wrap=4, xlabel=None,\n",
    "#                        rotation=45, hspace=1, top=0.95))\n",
    "kws = dict(dot=dict(figsize=(10, 5)), heat=dict(figsize=(15, 15)))\n",
    "_ = self.plot(kind=[\"matrix\", \"dot\"], genes=sig_contrast_genes,\n",
    "              standard_scale=\"obs\", col_celltype=col_batch,\n",
    "              by_group=col_celltype, **kws)\n",
    "\n",
    "# Results\n",
    "print(f\"\\n{'=' * 80}\\nSignificant Senescence Gene Contrasts\\n{'=' * 80}\\n\")\n",
    "print(sig_contrast_genes)\n",
    "res_snc_contrast"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fbaa031",
   "metadata": {},
   "source": [
    "### Snc versus Not (Group=Covariate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8e0b850",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group = Covariate\n",
    "out_edgr_contrasts_snc = scflow.ax.run_deg_edgr(\n",
    "    self.rna, \"Senescent_Cell_Label\",\n",
    "    col_covariate=col_batch, formula=None,\n",
    "    key_treatment=\"Senescent\", key_control=\"Non-Senescent\",\n",
    "    log2fc_thresh=0, n_top_vars=24, xlabel_rotation=45,\n",
    "    hspace=0.5, wspace=0.2, top=0.92, legend_loc=\"upper right\",\n",
    "    col_celltype=col_celltype, col_sample=col_sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a63d132",
   "metadata": {},
   "source": [
    "### SnC versus Not SnC (Group=Subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfb32a42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aged Subset\n",
    "subset_old = self.rna.obs[col_age] == keys[col_age][\"key_treatment\"]\n",
    "\n",
    "# Old Ground Control SnCs\n",
    "print(f\"\\n\\n{'=' * 80}\\nGround Control (Aged)\\n{'=' * 80}\\n\\n\")\n",
    "subset_gc = self.rna.obs[col_condition] == keys[\n",
    "    col_condition][\"key_control\"]\n",
    "out_edgr_contrasts_snc_gc = scflow.ax.run_deg_edgr(\n",
    "    self.rna[subset_gc & subset_old], \"Senescent_Cell_Label\",\n",
    "    key_treatment=\"Senescent\", top=0.92,\n",
    "    key_control=\"Non-Senescent\", log2fc_thresh=0, n_top_vars=24,\n",
    "    xlabel_rotation=45, hspace=0.5, wspace=0.2, legend_loc=\"upper right\",\n",
    "    col_celltype=col_celltype, col_sample=col_sample)\n",
    "if len(self.rna[subset_gc & subset_old].obs[col_batch].unique()) > 1:\n",
    "    raise ValueError(\"Check subsetting\")\n",
    "key_1 = self.rna[subset_gc & subset_old].obs[col_batch].unique()[0]\n",
    "\n",
    "# Old Spaceflight SnCs\n",
    "print(f\"\\n\\n{'=' * 80}\\nSpaceflight (Aged)\\n{'=' * 80}\\n\\n\")\n",
    "subset_sf = self.rna.obs[col_condition] == keys[\n",
    "    col_condition][\"key_treatment\"]\n",
    "out_edgr_contrasts_snc_sf = scflow.ax.run_deg_edgr(\n",
    "    self.rna[subset_sf & subset_old], \"Senescent_Cell_Label\",\n",
    "    key_treatment=\"Senescent\", key_control=\"Non-Senescent\",\n",
    "    log2fc_thresh=0, n_top_vars=24, xlabel_rotation=45,\n",
    "    hspace=0.5, wspace=0.2, top=0.92, legend_loc=\"upper right\",\n",
    "    col_celltype=col_celltype, col_sample=col_sample)\n",
    "if len(self.rna[subset_sf & subset_old].obs[col_batch].unique()) > 1:\n",
    "    raise ValueError(\"Check subsetting\")\n",
    "key_2 = self.rna[subset_sf & subset_old].obs[col_batch].unique()[0]\n",
    "\n",
    "# Compare\n",
    "res_snc_gc = out_edgr_contrasts_snc_gc[0][out_edgr_contrasts_snc_gc[\n",
    "    0].adj_p_value < 0.001].sort_values(\"adj_p_value\").set_index(\"variable\")\n",
    "res_snc_sf = out_edgr_contrasts_snc_sf[0][out_edgr_contrasts_snc_sf[\n",
    "    0].adj_p_value < 0.001].sort_values(\"adj_p_value\").set_index(\"variable\")\n",
    "res_snc_grps = pd.concat([\n",
    "    res_snc_gc, res_snc_sf], keys=[key_1, key_2], names=[col_condition])\n",
    "res_snc_grps = res_snc_grps[res_snc_grps.abs_log_fc >= 1]\n",
    "# top_tx_g_snc = res_snc_grps.loc[keys[col_condition][\n",
    "#     \"key_treatment\"]].sort_values(\"abs_log_fc\", ascending=False).index\n",
    "# print(res_snc_grps.unstack(0).replace(np.nan, \"\")[[\n",
    "#     \"log_fc\", \"adj_p_value\"]].loc[top_tx_g_snc])\n",
    "# just_sf = res_snc_grps.loc[keys[col_condition][\n",
    "#     \"key_treatment\"]].index.difference(res_snc_grps.loc[keys[\n",
    "#         col_condition][\"key_control\"]].index)\n",
    "# just_gc = res_snc_grps.loc[keys[col_condition][\n",
    "#     \"key_control\"]].index.difference(res_snc_grps.loc[keys[\n",
    "#         col_condition][\"key_treatment\"]].index)\n",
    "# print(res_snc_grps.loc[:, just_gc, :].sort_values(\n",
    "#     \"abs_log_fc\", ascending=False)[[\"log_fc\", \"adj_p_value\"]], \"\\n\\n\")\n",
    "# res_snc_grps.loc[:, just_sf, :].sort_values(\n",
    "#     \"abs_log_fc\", ascending=False)[[\"log_fc\", \"adj_p_value\"]]\n",
    "res_snc_grps.unstack(0).replace(np.nan, \"\")[[\"log_fc\", \"adj_p_value\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bf3e1d8",
   "metadata": {},
   "source": [
    "### Group DEGs (SnC=Covariate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e918a15d",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_edgr_contrasts_snc_cov = {}\n",
    "for a in self.rna.obs[col_age].unique():\n",
    "    print(f\"\\n\\n{'=' * 80}\\nAge = {a}\\n{'=' * 80}\\n\\n\")\n",
    "    out_edgr_contrasts_snc_cov[a] = scflow.ax.run_deg_edgr(\n",
    "        self.rna[self.rna.obs[col_age] == a], col_condition,\n",
    "        col_covariate=\"Senescent_Cell_Label\",\n",
    "        **keys[col_condition], log2fc_thresh=0, n_top_vars=24,\n",
    "        xlabel_rotation=45, hspace=0.5, wspace=0.2, top=0.85,\n",
    "        legend_loc=\"upper right\", col_celltype=col_celltype,\n",
    "        col_sample=col_sample)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4612d5e",
   "metadata": {},
   "source": [
    "### Group DEGs (SnC=Subset)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9075d404",
   "metadata": {},
   "source": [
    "##### All Groups SnCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79ffde96",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_edgr_groups_snc = scflow.ax.run_deg_edgr(\n",
    "    self.rna[(self.rna.obs[\"Senescent_Cell_Label\"] == \"Senescent\")],\n",
    "    col_batch, **keys[col_batch], col_sample=col_sample,\n",
    "    log2fc_thresh=0, n_top_vars=25, col_celltype=col_celltype)\n",
    "print(list(out_edgr_groups[0].variable[:25]))\n",
    "out_edgr_batches_snc = out_edgr_groups[0][(\n",
    "    out_edgr_groups[0].abs_log_fc > 1) & (\n",
    "        out_edgr_groups[0].adj_p_value < 0.001)].sort_values(\n",
    "            \"adj_p_value\", ascending=True)\n",
    "out_edgr_batches_top = out_edgr_batches.groupby(\"contrast\").apply(\n",
    "    lambda x: x.sort_values(\"adj_p_value\", ascending=True).iloc[\n",
    "        :20]).set_index(\"variable\", append=True).reset_index(\n",
    "            1, drop=True).rename_axis([col_condition, \"variable\"])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90b077d6",
   "metadata": {},
   "source": [
    "##### Just Aged SnCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "348a2c75",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_edgr_contrasts_sncto, snc_degs_by_condition_age_ct = {}, {}\n",
    "thresh_p_c, thresh_lfcabs_c = 0.05, 0.5  # thresholds\n",
    "subs_old = self.rna.obs[col_age] == keys[col_age][\"key_treatment\"]\n",
    "subs_sf = self.rna.obs[col_condition] == keys[col_condition][\"key_treatment\"]\n",
    "# subs_ac = (self.rna.obs[\"Senescent_Cell_Label\"] == \"Senescent\") & (\n",
    "#     subs_old | subs_sf)\n",
    "subs_ac = (self.rna.obs[\"Senescent_Cell_Label\"] == \"Senescent\") & (subs_old)\n",
    "out_edgr_contrasts_sncto[\"Overall\"] = scflow.ax.run_deg_edgr(\n",
    "        self.rna[subs_ac], col_condition,\n",
    "        # col_covariate=col_age,\n",
    "        **keys[col_condition],\n",
    "        log2fc_thresh=0, n_top_vars=24,\n",
    "        legend_loc=\"upper center\", fig_title=f\"Overall\",\n",
    "        wspace=0.25, bottom=0.3, hspace=1 if len(x) > 10 else 0.4, top=0.95,\n",
    "        kws_xticks=dict(fontsize=10, rotation=45 if len(x) > 10 else None),\n",
    "        col_celltype=col_celltype, col_sample=col_sample)\n",
    "plt.show()\n",
    "tmp = out_edgr_contrasts_sncto[\"Overall\"][0]\n",
    "tmp = tmp[(tmp.adj_p_value < thresh_p_c) & (\n",
    "    tmp.abs_log_fc >= thresh_lfcabs_c)]  # filter by p & lfc\n",
    "snc_degs_by_condition_age_ct[\"Overall\"] = list(\n",
    "    tmp.sort_values(\"adj_p_value\", ascending=True).variable)\n",
    "for x in self.rna.obs[col_celltype].unique():\n",
    "    print(x)\n",
    "    out_edgr_contrasts_sncto[x] = scflow.ax.run_deg_edgr(\n",
    "        self.rna[(subs_ac) & (self.rna.obs[col_celltype] == x)],\n",
    "        col_condition,\n",
    "        # col_covariate=col_age,\n",
    "        **keys[col_condition],\n",
    "        log2fc_thresh=0, n_top_vars=24,\n",
    "        legend_loc=\"upper center\", fig_title=f\"Senescent {x}\",\n",
    "        wspace=0.25, bottom=0.3, hspace=1 if len(x) > 10 else 0.4, top=0.95,\n",
    "        kws_xticks=dict(fontsize=10, rotation=45 if len(x) > 10 else None),\n",
    "        col_celltype=col_celltype, col_sample=col_sample)\n",
    "    plt.show()\n",
    "    tmp = out_edgr_contrasts_sncto[x][0]\n",
    "    tmp = tmp[(tmp.adj_p_value < thresh_p_c) & (\n",
    "        tmp.abs_log_fc >= thresh_lfcabs_c)]  # filter by p & lfc\n",
    "    snc_degs_by_condition_age_ct[x] = list(tmp.sort_values(\n",
    "            \"adj_p_value\", ascending=True).variable)\n",
    "    del tmp\n",
    "out_edgr_contrasts_snc_all_top = pd.concat([out_edgr_contrasts_sncto[x][\n",
    "    0].set_index(\"variable\").loc[snc_degs_by_condition_age_ct[\n",
    "        x]] for x in out_edgr_contrasts_sncto], keys=out_edgr_contrasts_sncto,\n",
    "                                           names=[col_celltype])\n",
    "out_edgr_contrasts_snc_all = pd.concat([\n",
    "    out_edgr_contrasts_sncto[x][0].set_index(\"variable\")\n",
    "    for x in out_edgr_contrasts_sncto], keys=out_edgr_contrasts_sncto,\n",
    "                                       names=[col_celltype])\n",
    "out_edgr_contrasts_snc_all_top"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "305886bb",
   "metadata": {},
   "source": [
    "### More Specific Contrasts (`pyDESeq2`)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fead6cd1",
   "metadata": {},
   "source": [
    "#### Group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5e784ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pdata =  scflow.tl.create_pseudobulk(\n",
    "#     self.rna, [col_sample, col_condition, col_age],\n",
    "#     col_celltype, layer=\"counts\", mode=\"sum\")\n",
    "# design = f\"~{col_condition} + {col_age} + {col_condition} * {col_age}\"\n",
    "# # edgr = pt.tl.EdgeR(pdata, design=design)\n",
    "# pds2 = pt.tl.PyDESeq2(adata=pdata, design=design)\n",
    "# pds2.fit()\n",
    "# key_control, key_treatment = [keys[col_condition][k] for k in [\n",
    "#     \"key_control\", \"key_treatment\"]]\n",
    "# key_control_cov, key_treatment_cov = [keys[col_age][k] for k in [\n",
    "#     \"key_control\", \"key_treatment\"]]\n",
    "# sf_old = {col_condition: key_treatment, col_age: key_treatment_cov}\n",
    "# sf_yng = {col_condition: key_treatment, col_age: key_control_cov}\n",
    "# gc_old = {col_condition: key_control, col_age: key_treatment_cov}\n",
    "# gc_yng = {col_condition: key_control, col_age: key_control_cov}\n",
    "\n",
    "# # Spaceflight-Specific Age DEGs\n",
    "# spaceflight_specfic_age_degs = (pds2.cond(**sf_old) - pds2.cond(\n",
    "#     **sf_yng)) - (pds2.cond(**gc_old) - pds2.cond(**gc_yng))\n",
    "# interaction_res_df = pds2.test_contrasts(spaceflight_specfic_age_degs)\n",
    "# pds2.plot_volcano(interaction_res_df, log2fc_thresh=0)\n",
    "\n",
    "# # Age-Specific Spaceflight DEGs\n",
    "# age_specfic_sf_degs = (pds2.cond(**sf_old) - pds2.cond(**gc_old)) - (\n",
    "#     pds2.cond(**sf_yng) - pds2.cond(**gc_yng))\n",
    "# interaction_res_df2 = pds2.test_contrasts(age_specfic_sf_degs)\n",
    "# pds2.plot_volcano(interaction_res_df2, log2fc_thresh=0)\n",
    "\n",
    "# # Old vs. Young Spaceflight\n",
    "# age_sf_ix = pds2.cond(**sf_old) - pds2.cond(**sf_yng)\n",
    "# interaction_res_df3 = pds2.test_contrasts(age_sf_ix)\n",
    "# pds2.plot_volcano(interaction_res_df3, log2fc_thresh=0)\n",
    "\n",
    "# # Old vs. Young GC\n",
    "# age_gc_ix = pds2.cond(**gc_old) - pds2.cond(**gc_yng)\n",
    "# interaction_res_df4 = pds2.test_contrasts(age_gc_ix)\n",
    "# pds2.plot_volcano(interaction_res_df4, log2fc_thresh=0)\n",
    "\n",
    "# # SF vs. GC Old\n",
    "# cond_old_ix = pds2.cond(**sf_old) - pds2.cond(**gc_old)\n",
    "# interaction_res_df5 = pds2.test_contrasts(cond_old_ix)\n",
    "# pds2.plot_volcano(interaction_res_df5, log2fc_thresh=0)\n",
    "\n",
    "# # SF vs. GC Young\n",
    "# cond_yng_ix = pds2.cond(**sf_yng) - pds2.cond(**gc_yng)\n",
    "# interaction_res_df6 = pds2.test_contrasts(cond_yng_ix)\n",
    "# pds2.plot_volcano(interaction_res_df6, log2fc_thresh=0)\n",
    "\n",
    "# # Compare\n",
    "# thresh_lfc, thresh_p = 1, 0.001\n",
    "# sf_deg_res_df_old = interaction_res_df5[(\n",
    "#     interaction_res_df5.adj_p_value < thresh_p) & (\n",
    "#         interaction_res_df5.log_fc.abs() >= thresh_lfc)].set_index(\n",
    "#             \"variable\")\n",
    "# sf_deg_res_df_yng = interaction_res_df6[(\n",
    "#     interaction_res_df6.adj_p_value < thresh_p) & (\n",
    "#         interaction_res_df6.log_fc.abs() >= thresh_lfc)].set_index(\n",
    "#             \"variable\")\n",
    "# sf_deg_res = pd.concat([sf_deg_res_df_old, sf_deg_res_df_yng],\n",
    "#                        keys=[\"Old\", \"Young\"], names=[\"Age Group\"])\n",
    "# sf_deg_res.loc[:, \"log_fc_abs\"] = sf_deg_res.log_fc.abs()\n",
    "# sf_deg_res = sf_deg_res.sort_values(\"log_fc_abs\", ascending=False)[[\n",
    "#     \"log_fc\", \"adj_p_value\"]].unstack(0)\n",
    "# print(\"SF DEGs in Both Young & Old: \", sf_deg_res.dropna().index.values)\n",
    "# print(\"\\n\\nSF DEGs in Just Young\", list(sf_deg_res_df_yng.index.difference(\n",
    "#     sf_deg_res_df_old.index)), \"\\n\\n\")\n",
    "# print(\"SF DEGs in Just Old\", list(sf_deg_res_df_old.index.difference(\n",
    "#     sf_deg_res_df_yng.index)), \"\\n\\n\")\n",
    "# sf_deg_res.replace(np.nan, \"\")\n",
    "\n",
    "# # Marker Set Overlap\n",
    "# print(\"Both: \", marker_gene_sets.apply(\n",
    "#     lambda i: set(sf_deg_res.dropna().index).intersection(set(i))).apply(\n",
    "#         lambda x: x if len(x) > 0 else np.nan).dropna(), \"\\n\\n\")\n",
    "# print(\"Young: \", marker_gene_sets.apply(\n",
    "#     lambda i: sf_deg_res_df_yng.index.difference(\n",
    "#         sf_deg_res_df_old.index).intersection(set(i))).apply(\n",
    "#         lambda x: x if len(x) > 0 else np.nan).dropna(), \"\\n\\n\")\n",
    "# print(\"Old: \", marker_gene_sets.apply(\n",
    "#     lambda i: sf_deg_res_df_old.index.difference(\n",
    "#         sf_deg_res_df_yng.index).intersection(set(i))).apply(\n",
    "#         lambda x: x if len(x) > 0 else np.nan).dropna(), \"\\n\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75aad858",
   "metadata": {},
   "source": [
    "#### SnCs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6604afaf",
   "metadata": {},
   "source": [
    "##### Old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecc5769c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup\n",
    "design = (f\"~{col_condition} + Senescent_Cell_Label + \"\n",
    "          f\"{col_condition} * Senescent_Cell_Label\")\n",
    "pdata2 =  scflow.tl.create_pseudobulk(\n",
    "    self.rna[self.rna.obs[col_age] == keys[col_age][\"key_treatment\"]], [\n",
    "        col_sample, col_condition, \"Senescent_Cell_Label\"],\n",
    "    \"Senescent_Cell_Label_by_Type\", layer=\"counts\", mode=\"sum\")\n",
    "pds2_snc_old = pt.tl.PyDESeq2(adata=pdata2, design=design)\n",
    "pds2_snc_old.fit()\n",
    "key_control, key_treatment = [keys[col_condition][k] for k in [\n",
    "    \"key_control\", \"key_treatment\"]]\n",
    "key_control_cov, key_treatment_cov = \"Non-Senescent\", \"Senescent\"\n",
    "sf_sc = {col_condition: key_treatment,\n",
    "         \"Senescent_Cell_Label\": key_treatment_cov}\n",
    "sf_ns = {col_condition: key_treatment,\n",
    "         \"Senescent_Cell_Label\": key_control_cov}\n",
    "gc_snc = {col_condition: key_control,\n",
    "          \"Senescent_Cell_Label\": key_treatment_cov}\n",
    "gc_ns = {col_condition: key_control, \"Senescent_Cell_Label\": key_control_cov}\n",
    "\n",
    "# SnC DEGs\n",
    "res_df_snc = pds2_snc_old.test_contrasts(pds2_snc_old.contrast(\n",
    "    column=\"Senescent_Cell_Label\", baseline=key_control_cov,\n",
    "    group_to_compare=key_treatment_cov))\n",
    "print(res_df_snc[res_df_snc.adj_p_value < 0.001].head(10))\n",
    "pds2_snc_old.plot_volcano(res_df_snc, log2fc_thresh=0)\n",
    "\n",
    "# Spaceflight-Specific SnC DEGs\n",
    "spaceflight_specfic_snc_degs = (pds2_snc_old.cond(\n",
    "    **sf_sc) - pds2_snc_old.cond(**sf_ns)) - (\n",
    "        pds2_snc_old.cond(**gc_snc) - pds2_snc_old.cond(**gc_ns))\n",
    "interaction_res_df_snc_sf = pds2_snc_old.test_contrasts(\n",
    "    spaceflight_specfic_snc_degs)\n",
    "print(interaction_res_df_snc_sf[\n",
    "    interaction_res_df_snc_sf.adj_p_value < 0.001].head(10))\n",
    "pds2_snc_old.plot_volcano(interaction_res_df_snc_sf, log2fc_thresh=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e026c7aa",
   "metadata": {},
   "source": [
    "##### Young"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0ca7bc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup\n",
    "design = (f\"~{col_condition} + Senescent_Cell_Label + \"\n",
    "          f\"{col_condition} * Senescent_Cell_Label\")\n",
    "pdata2 =  scflow.tl.create_pseudobulk(\n",
    "    self.rna[self.rna.obs[col_age] == keys[col_age][\"key_treatment\"]], [\n",
    "        col_sample, col_condition, \"Senescent_Cell_Label\"],\n",
    "    \"Senescent_Cell_Label_by_Type\", layer=\"counts\", mode=\"sum\")\n",
    "pds2_snc_yng = pt.tl.PyDESeq2(adata=pdata2, design=design)\n",
    "pds2_snc_yng.fit()\n",
    "key_control, key_treatment = [keys[col_condition][k] for k in [\n",
    "    \"key_control\", \"key_treatment\"]]\n",
    "key_control_cov, key_treatment_cov = \"Non-Senescent\", \"Senescent\"\n",
    "sf_sc = {col_condition: key_treatment,\n",
    "         \"Senescent_Cell_Label\": key_treatment_cov}\n",
    "sf_ns = {col_condition: key_treatment,\n",
    "         \"Senescent_Cell_Label\": key_control_cov}\n",
    "gc_snc = {col_condition: key_control,\n",
    "          \"Senescent_Cell_Label\": key_treatment_cov}\n",
    "gc_ns = {col_condition: key_control, \"Senescent_Cell_Label\": key_control_cov}\n",
    "\n",
    "# SnC DEGs\n",
    "res_df_snc = pds2_snc_yng.test_contrasts(pds2_snc_yng.contrast(\n",
    "    column=\"Senescent_Cell_Label\", baseline=key_control_cov,\n",
    "    group_to_compare=key_treatment_cov))\n",
    "print(res_df_snc[res_df_snc.adj_p_value < 0.001].head(10))\n",
    "pds2_snc_yng.plot_volcano(res_df_snc, log2fc_thresh=0)\n",
    "\n",
    "# Spaceflight-Specific SnC DEGs\n",
    "spaceflight_specfic_snc_degs = (pds2_snc_yng.cond(\n",
    "    **sf_sc) - pds2_snc_yng.cond(**sf_ns)) - (\n",
    "        pds2_snc_yng.cond(**gc_snc) - pds2_snc_yng.cond(**gc_ns))\n",
    "interaction_res_df_snc_sf = pds2_snc_yng.test_contrasts(\n",
    "    spaceflight_specfic_snc_degs)\n",
    "pds2_snc_yng.plot_volcano(interaction_res_df_snc_sf, log2fc_thresh=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc67c178",
   "metadata": {},
   "source": [
    "#### SnCs (More Permissive Threshold)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d2922db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare Less & More Permissive Thresholds\n",
    "print(\"Less & More Permissive SnC Threshold: \", self.rna.obs[[\n",
    "    \"Senescent_Cell_Label\", f\"Senescent_Cell_Label_{use_metric}_{p_h}\"\n",
    "    ]].value_counts())\n",
    "sen_lab = f\"Senescent_Cell_Label_{use_metric}_{p_h}\"\n",
    "\n",
    "# Setup\n",
    "design = (f\"~{col_condition} + {sen_lab} + \"\n",
    "          f\"{col_condition} * {sen_lab}\")\n",
    "pdata2_3 =  scflow.tl.create_pseudobulk(\n",
    "    self.rna[self.rna.obs[col_age] == keys[col_age][\"key_treatment\"]], [\n",
    "        col_sample, col_condition, sen_lab],\n",
    "    col_celltype, layer=\"counts\", mode=\"sum\")\n",
    "pds2_snc_old_3 = pt.tl.PyDESeq2(adata=pdata2_3, design=design)\n",
    "pds2_snc_old_3.fit()\n",
    "key_control, key_treatment = [keys[col_condition][k] for k in [\n",
    "    \"key_control\", \"key_treatment\"]]\n",
    "key_control_cov, key_treatment_cov = \"Non-Senescent\", \"Senescent\"\n",
    "sf_sc = {col_condition: key_treatment, sen_lab: key_treatment_cov}\n",
    "sf_ns = {col_condition: key_treatment, sen_lab: key_control_cov}\n",
    "gc_snc = {col_condition: key_control, sen_lab: key_treatment_cov}\n",
    "gc_ns = {col_condition: key_control, sen_lab: key_control_cov}\n",
    "\n",
    "# SnC DEGs\n",
    "res_df_snc_3 = pds2_snc_old_3.test_contrasts(pds2_snc_old_3.contrast(\n",
    "    column=sen_lab, baseline=key_control_cov,\n",
    "    group_to_compare=key_treatment_cov))\n",
    "print(res_df_snc_3[res_df_snc_3.adj_p_value < 0.001].head(10))\n",
    "pds2_snc_old_3.plot_volcano(res_df_snc_3, log2fc_thresh=0)\n",
    "\n",
    "# Spaceflight-Specific SnC DEGs\n",
    "spaceflight_specfic_snc_degs_3 = (pds2_snc_old_3.cond(\n",
    "    **sf_sc) - pds2_snc_old_3.cond(**sf_ns)) - (\n",
    "        pds2_snc_old_3.cond(**gc_snc) - pds2_snc_old_3.cond(**gc_ns))\n",
    "interaction_res_df_snc_sf_3 = pds2_snc_old_3.test_contrasts(\n",
    "    spaceflight_specfic_snc_degs_3)\n",
    "pds2_snc_old_3.plot_volcano(interaction_res_df_snc_sf_3, log2fc_thresh=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01327816",
   "metadata": {},
   "source": [
    "### MAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d760bba4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R\n",
    "# find_de_MAST_RE <- function(adata_, formula){\n",
    "#     # create a MAST object\n",
    "#     sca <- SceToSingleCellAssay(adata_, class=\"SingleCellAssay\")\n",
    "#     print(\"Dimensions before subsetting:\")\n",
    "#     print(dim(sca))\n",
    "#     print(\"\")\n",
    "#     # keep genes that are expressed in more than 10% of all cells\n",
    "#     sca <- sca[freq(sca) > 0.1,]\n",
    "#     print(\"Dimensions after subsetting:\")\n",
    "#     print(dim(sca), \"\\n\")\n",
    "#     cdr2 <- colSums(assay(sca) > 0)\n",
    "#     colData(sca)$ngeneson <- scale(cdr2)\n",
    "#     # store the columns that we are interested in as factors\n",
    "#     label <- factor(colData(sca)$label)\n",
    "#     # set the reference level\n",
    "#     label <- relevel(label,\"ctrl\")\n",
    "#     colData(sca)$label <- label\n",
    "#     celltype <- factor(colData(sca)$cell_type)\n",
    "#     colData(sca)$celltype <- celltype\n",
    "#     # same for donors (which we need to model random effects)\n",
    "#     replicate <- factor(colData(sca)$replicate)\n",
    "#     colData(sca)$replicate <- replicate\n",
    "#     # create a group per condition-celltype combination\n",
    "#     colData(sca)$group <- paste0(colData(\n",
    "#         adata_)$label, \".\", colData(adata_)$cell_type)\n",
    "#     colData(sca)$group <- factor(colData(sca)$group)\n",
    "#     # define and fit the model\n",
    "#     zlmCond <- zlm(formula=formula,\n",
    "#                    sca=sca,\n",
    "#                    method=\"glmer\",\n",
    "#                    ebayes=F,\n",
    "#                    strictConvergence=F,\n",
    "#                    fitArgsD=list(nAGQ = 0)) # to speed up calculations\n",
    "#     # perform likelihood-ratio test for the condition\n",
    "#     summaryCond <- summary(zlmCond, doLRT=\"groupstim.CD14_Monocytes\")\n",
    "#     # get the table with log-fold changes and p-values\n",
    "#     summaryDt <- summaryCond$datatable\n",
    "#     return(summaryDt)\n",
    "# }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "458a366e",
   "metadata": {},
   "source": [
    "# SnC GEX Profiles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed4a4d4e",
   "metadata": {},
   "source": [
    "## Pathway Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4958f120",
   "metadata": {},
   "outputs": [],
   "source": [
    "names = gp.get_library_name()\n",
    "# perts_geo = [\"Aging\", \"Drug\", \"Disease\"]\n",
    "perts_geo = [\"Aging\", \"Disease\"]\n",
    "other_sets = [\"GTEx_Aging_Signatures_2021\", f\"HDSigDB_{species}_2021\",\n",
    "              \"MSigDB_Hallmark_2020\",  # \"DGIdb_Drug_Targets_2024\",\n",
    "              \"MSigDB_Oncogenic_Signatures\", \"OMIM_Disease\",\n",
    "              # \"Reimport pdb;actome_Pathways_2024\", \"Reactome_2022\",\n",
    "              f\"WikiPathways_2024_{species}\", f\"WikiPathways_2019_{species}\"]\n",
    "other_sets += [i for i in names if \"GO_Biological_Process\" in i]\n",
    "gene_sets = dict(zip([\"up\", \"down\"], [[\n",
    "    f\"{u}_Perturbations_from_GEO_{i}\" for u in perts_geo] + [\n",
    "        f\"Disease_Signatures_from_GEO_{i}_2014\"] + other_sets\n",
    "    for i in [\"up\", \"down\"]]))\n",
    "\n",
    "\n",
    "def fx_replace_string(string):\n",
    "    \"\"\"Make short version of pathway terms.\"\"\"\n",
    "    string = re.sub(\"aging:([0-9]+)\", \"aging\", re.sub(\n",
    "        \"([0-9]+) months v ([0-9]+) mo(nths)?\", \"\\\\1 v \\\\2 mo.\", re.sub(\n",
    "            \"SRP[0-9]+ \", \"\", re.sub(\"GSE[0-9]+ \", \"\", string))))\n",
    "    string = re.sub(\".*.xlsx.\", \"\", re.sub(\n",
    "        \".*Supplementary Data ..\", \"\", re.sub(\n",
    "            \"PMID([0-9])+\", \"\", re.sub(\"([0-9]+) years v ([0-9]+) years\",\n",
    "                                       \"\\\\1 v \\\\2 yrs.\", string))))\n",
    "    string = re.sub(\"Expression Of \", \"\", re.sub(\".*.xls\", \"\", re.sub(\n",
    "        \".*.XLSX\", \"\", re.sub(\n",
    "            \".*Supplementary Table [0-9]+[-]?\", \"\", string))))\n",
    "    return string"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbd8ea6b",
   "metadata": {},
   "source": [
    "### Endrichr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d962f710",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_pathway_snc_edgr_contrasts = scflow.ax.run_enrichr(\n",
    "    out_edgr_contrasts_snc_all_top, col_grouping=[\"Comparison\", col_celltype],\n",
    "    gene_sets=gene_sets, fx_replace_string=fx_replace_string, title=\"Overall\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51d35361",
   "metadata": {},
   "source": [
    "### SnC vs. Not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24555721",
   "metadata": {},
   "outputs": [],
   "source": [
    "# res_dc_ulm = scflow.ax.run_decoupler_ulm(\n",
    "#         self.rna[self.rna.obs[\"Senescent_Cell_Label\"] == \"Senescent\"],\n",
    "#         # \"Senescent_Cell_Label_by_Type\",\n",
    "#         # \"Senescent_Cell_Label\",\n",
    "#         col_condition, col_condition=\"Senescent_Cell_Label_by_Type\",\n",
    "#         species=species, resource=\"hallmark\",\n",
    "#         top=0.95, hspace=1, wspace=0.5, figsize=(15, 10))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53ed84c7",
   "metadata": {},
   "source": [
    "#### Pathway Set = Progeny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b4a8484",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dc_ulm = scflow.ax.run_decoupler_ulm(\n",
    "        self.rna[self.rna.obs[\"Senescent_Cell_Label\"] == \"Senescent\"],\n",
    "        col_condition, col_condition=\"Senescent_Cell_Label_by_Type\",\n",
    "        species=species, resource=\"progeny\",\n",
    "        hspace=0.5, wspace=0.5, top=0.95, figsize=(15, 15))\n",
    "res_dc_ulm_scores = pd.concat([res_dc_ulm[0][x][1].set_index(\n",
    "    \"group\").rename_axis(col_condition).set_index(\n",
    "        \"name\", append=True) for x in res_dc_ulm[0]],\n",
    "                              keys=res_dc_ulm[0], names=[col_celltype])\n",
    "res_dc_ulm_pathways = pd.concat([pd.Series(res_dc_ulm[0][x][2])\n",
    "                                 for x in res_dc_ulm[0]], keys=res_dc_ulm[0])\n",
    "print(res_dc_ulm_pathways)\n",
    "res_dc_ulm_scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8dd102b",
   "metadata": {},
   "source": [
    "#### Pathway Set = Hallmark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a173e33",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dc_ulm_hm = scflow.ax.run_decoupler_ulm(\n",
    "        self.rna[self.rna.obs[\"Senescent_Cell_Label\"] == \"Senescent\"],\n",
    "        col_condition, col_condition=\"Senescent_Cell_Label_by_Type\",\n",
    "        species=species, resource=\"hallmark\",\n",
    "        hspace=0.5, top=0.95, figsize=(15, 15))\n",
    "res_dc_ulm_hm_scores = pd.concat([res_dc_ulm_hm[0][x][1].set_index(\n",
    "    \"group\").rename_axis(col_condition).set_index(\n",
    "        \"name\", append=True) for x in res_dc_ulm_hm[0]],\n",
    "                                 keys=res_dc_ulm_hm[0], names=[col_celltype])\n",
    "res_dc_ulm_hm_scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d0360f0",
   "metadata": {},
   "source": [
    "### Individual Cell Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "392e0a80",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_aucell = scflow.ax.run_decoupler_aucell(\n",
    "    self.rna, col_covariates=[\"Senescent_Cell_Label_by_Type\", col_condition],\n",
    "    resource=\"progeny\", species=species, inplace=True)\n",
    "_ = self.plot(genes=list(self.rna.obsm[\"score_aucell\"].columns),\n",
    "              subset=self.rna.obs[\"Senescent_Cell\"],\n",
    "              col_celltype=col_condition, kind=\"violin\",\n",
    "              by_group=\"Senescent_Cell_Label_by_Type\", figsize=(15, 15))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1704aea",
   "metadata": {},
   "source": [
    "#### DEGs (`find_markers`) by Group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4f02ce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_pathway_snc_degs = scflow.ax.run_enrichr(\n",
    "    df_snc_markers_top, col_grouping=[\"Comparison\", col_celltype],\n",
    "    gene_sets=gene_sets, fx_replace_string=fx_replace_string, title=\"Overall\")\n",
    "out_pathway_snc_degs_gc = scflow.ax.run_enrichr(\n",
    "    df_snc_markers_top_gc, col_grouping=[\"Comparison\", col_celltype],\n",
    "    gene_sets=gene_sets, fx_replace_string=fx_replace_string,\n",
    "    title=keys[col_condition][\"key_control\"])\n",
    "out_pathway_snc_degs_sf = scflow.ax.run_enrichr(\n",
    "    df_snc_markers_top_sf, col_grouping=[\"Comparison\", col_celltype],\n",
    "    gene_sets=gene_sets, fx_replace_string=fx_replace_string,\n",
    "    title=keys[col_condition][\"key_treatment\"])\n",
    "# out_pathway_snc_degs[1].drop(\"All\", level=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "560593c9",
   "metadata": {},
   "source": [
    "#### Edger Contrasts (Pseudo-Bulk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fb9bf51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GSEA\n",
    "print(\"\\n\\nGSEA Enrichment Analysis\\n\\n\")\n",
    "out_pathway_snc_degs = scflow.ax.run_enrichr(\n",
    "    out_edgr_contrasts_snc_all_top, col_grouping=col_celltype,\n",
    "    gene_sets=gene_sets, fx_replace_string=fx_replace_string)\n",
    "\n",
    "# ToppGene\n",
    "print(\"\\n\\nToppGene Enrichment Analysis\\n\\n\")\n",
    "mks_ctrsts_snc = dict(out_edgr_contrasts_snc_all_top.reset_index(\n",
    "    1).variable.groupby(col_celltype).apply(lambda x: list(x)))  # SnC DEGs\n",
    "results_toppgene_degs = scflow.pp.annotate_by_toppgene(\n",
    "    mks_ctrsts_snc, species=species, min_genes=1,\n",
    "    categories=[\"Pathway\", \"PubMed\"])\n",
    "results_toppgene_degs = results_toppgene_degs.join(pd.Series(\n",
    "    mks_ctrsts_snc).to_frame(\"DEGs\").rename_axis(\"Gene Set\"))\n",
    "results_toppgene_degs = results_toppgene_degs.groupby(\"Gene Set\").apply(\n",
    "    lambda x: x.sort_values(\"PValue\", ascending=True), include_groups=False)\n",
    "pthwys = dict(results_toppgene_degs.groupby(\"Gene Set\").apply(\n",
    "    lambda x: [i.capitalize() for i in x.Name.unique()]))\n",
    "print(\"\\n\\n\".join([f\"{k}: {pthwys[k]}\" for k in pthwys]))\n",
    "results_toppgene_degs[[\"Name\", \"Symbols\", \"GenesInTerm\", \"GenesInQuery\",\n",
    "                       \"GenesInTermInQuery\", \"PValue\"]].reset_index(\n",
    "                           2, drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68b955c3",
   "metadata": {},
   "source": [
    "## Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65e3425c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Top 10 Senescence Genes or All?\n",
    "ggg = top_genes\n",
    "# ggg = genes\n",
    "\n",
    "# Mask for High Senescence Subset & Helper Variable for Figure Dimensions\n",
    "subset = self.rna.obs[\"Senescent_Cell\"].isin([1, True])\n",
    "num_cts = len(self.rna[subset].obs[col_celltype].unique())\n",
    "\n",
    "# Plot SnC Markers Overlap with Marker Sets\n",
    "sc.pl.violin(self.rna[subset], set(genes).intersection(\n",
    "    marker_gene_sets.explode().unique()),\n",
    "             groupby=col_batch, rotation=90)  # plot overlap w/ SnC markers\n",
    "\n",
    "# Plot Most Highly-Expressed Genes in Each Senescent Cell Type\n",
    "_ = self.plot(kind=[\"matrix\", \"heat\"], genes=top_genes,\n",
    "              col_celltype=col_celltype, layer=\"log1p\",\n",
    "              subset=subset, dendrogram=True, swap_axes=False, heat=dict(\n",
    "                  dendrogram=False), standard_scale=\"var\",\n",
    "              figsize=(len(top_genes) / 1.5, num_cts * 1.5))\n",
    "\n",
    "# Grouped by Top Cell Types\n",
    "_ = self.plot(kind=[\"matrix\", \"heat\"], genes=top_genes_dict,\n",
    "              standard_scale=\"obs\", col_celltype=col_celltype,\n",
    "              layer=\"log1p\", swap_axes=False, figsize=(len(ggg), num_cts),\n",
    "              subset=subset, dendrogram=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c284f5b",
   "metadata": {},
   "source": [
    "# Write Object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "344047b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Send Email with Output When Done\n",
    "if email is not None and html_out is not None:\n",
    "    os.system(f\"jupyter nbconvert --to html {cur_file}\")\n",
    "    os.system(f\"echo 'yay' | mutt -s 'JOB DONE' -a {html_out} -- {email}\")\n",
    "\n",
    "if overwrite is True or os.path.exists() is False:\n",
    "    self.rna.write_h5ad(file_path_new)\n",
    "else:\n",
    "    print(f\"\\n\\n\\n{'=' * 120}\\nOVERWRITE IS FALSE...\"\n",
    "          f\"NOT WRITING OBJECT!!!\\n{'=' * 120}\\n\\n\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c84e453",
   "metadata": {},
   "source": [
    "# SCRATCH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b106dbd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_degs = out_edgr_old_sf_v_gc.copy()\n",
    "# df_degs = out_edgr_batches_snc_top.copy()\n",
    "# df_degs = res_snc_grps.copy()\n",
    "df_degs = df_snc_markers.drop(\"All\").reset_index(0, drop=True)\n",
    "\n",
    "factor_name = col_celltype\n",
    "# factor_name = col_condition\n",
    "\n",
    "names = gp.get_library_name()\n",
    "# [i for i in names if \"brain\" in i.lower() or \"mouse\" in i.lower()]\n",
    "pathways_snc_vs_no, res_pathways_snc_vs_no = {}, {}\n",
    "thresh_p_path = 0.05\n",
    "# perts_geo = [\"Aging\", \"Drug\", \"Disease\"]\n",
    "perts_geo = [\"Aging\", \"Disease\"]\n",
    "other_sets = [\"GTEx_Aging_Signatures_2021\", f\"HDSigDB_{species}_2021\",\n",
    "              # \"DGIdb_Drug_Targets_2024\",\n",
    "              \"MSigDB_Hallmark_2020\",\n",
    "              \"MSigDB_Oncogenic_Signatures\", \"OMIM_Disease\",\n",
    "              # \"Reactome_Pathways_2024\", \"Reactome_2022\",\n",
    "              f\"WikiPathways_2024_{species}\", f\"WikiPathways_2019_{species}\"]\n",
    "other_sets += [i for i in names if \"GO_Biological_Process\" in i]\n",
    "gene_sets = dict(zip([\"up\", \"down\"], [[\n",
    "    f\"{u}_Perturbations_from_GEO_{i}\" for u in perts_geo] + [\n",
    "        f\"Disease_Signatures_from_GEO_{i}_2014\"] + other_sets\n",
    "    for i in [\"up\", \"down\"]]))\n",
    "# if species.lower() == \"mouse\":\n",
    "#     other_sets += [\"PerturbAtlas_MouseGenePerturbationSigs\"]\n",
    "for x in df_degs.reset_index()[factor_name].unique():\n",
    "    pathways_snc_vs_no[x], res_pathways_snc_vs_no[x] = {}, {}\n",
    "    for i in [\"up\", \"down\"]:\n",
    "        degs_tmp = df_degs.loc[x]\n",
    "        if i == \"up\":\n",
    "            degs_tmp = list(degs_tmp[degs_tmp.log_fc > 0].index.values)\n",
    "        else:\n",
    "            degs_tmp = list(degs_tmp[degs_tmp.log_fc < 0].index.values)\n",
    "        if len(degs_tmp) > 0:\n",
    "            g_s = [f\"{u}_Perturbations_from_GEO_{i}\" for u in perts_geo] + [\n",
    "                f\"Disease_Signatures_from_GEO_{i}_2014\"] + other_sets\n",
    "            g_s = [g for g in g_s if g in names]\n",
    "            pathways_snc_vs_no[x][i] = gp.enrichr(\n",
    "                gene_list=degs_tmp, gene_sets=g_s,\n",
    "                organism=species, cutoff=0.5)\n",
    "            res_pathways_snc_vs_no[x][i] = pathways_snc_vs_no[x][i].results[\n",
    "                pathways_snc_vs_no[x][i].results[\n",
    "                    \"Adjusted P-value\"] < thresh_p_path]\n",
    "    res_pathways_snc_vs_no[x] = pd.concat(res_pathways_snc_vs_no[x],\n",
    "                                          names=[\"Direction\"])\n",
    "res_pathways_snc_vs_no = pd.concat(res_pathways_snc_vs_no,\n",
    "                                   names=[col_celltype])\n",
    "res_pathways_snc_vs_no.loc[\n",
    "    :, \"Term Short\"] = res_pathways_snc_vs_no.Term.apply(\n",
    "        lambda x: re.sub(\"Expression Of \", \"\", re.sub(\".*.xls\", \"\", re.sub(\n",
    "            \".*.XLSX\", \"\",\n",
    "            re.sub(\".*Supplementary Table [0-9]+[-]?\", \"\", re.sub(\n",
    "                \".*.xlsx.\", \"\", re.sub(\".*Supplementary Data ..\", \"\", re.sub(\n",
    "                    \"PMID([0-9])+\", \"\", re.sub(\n",
    "                        \"([0-9]+) years v ([0-9]+) years\", \"\\\\1 v \\\\2 yrs.\",\n",
    "                        re.sub(\"aging:([0-9]+)\", \"aging\", re.sub(\n",
    "                            \"([0-9]+) months v ([0-9]+) mo(nths)?\",\n",
    "                            \"\\\\1 v \\\\2 mo.\", re.sub(\"SRP[0-9]+ \", \"\", re.sub(\n",
    "                                \"GSE[0-9]+ \", \"\", x)))))))))))))\n",
    "cts_i = res_pathways_snc_vs_no.reset_index()[col_celltype].unique()\n",
    "res_pathways_snc_top_vs_no = res_pathways_snc_vs_no.reset_index().groupby([\n",
    "    col_celltype, \"Direction\"]).apply(lambda x: x.sort_values(\n",
    "        \"Adjusted P-value\", ascending=True).head(10), include_groups=False)\n",
    "fig, axes = plt.subplots(*scflow.pl.square_grid(cts_i), figsize=(\n",
    "    20, 25), gridspec_kw=dict(wspace=1.5, top=0.95), squeeze=False)\n",
    "for i, x in enumerate(cts_i):\n",
    "    gp.dotplot(\n",
    "        res_pathways_snc_top_vs_no.loc[x].reset_index(),\n",
    "        column=\"Adjusted P-value\",\n",
    "        # column=\"Overlap\",\n",
    "        x=\"Direction\", y=\"Term Short\",\n",
    "        ax=axes.flatten()[i], size=20, top_term=100,\n",
    "        title=x, xticklabels_rot=45, yticklabels_rot=45,\n",
    "        show_ring=True, marker=\"o\", cutoff=0.5)\n",
    "    axes.flatten()[i].tick_params(axis=\"y\", labelsize=8,\n",
    "                                  labelfontfamily=\"serif\")\n",
    "    axes.flatten()[i].tick_params(axis=\"x\", labelsize=10,\n",
    "                                  labelfontfamily=\"serif\")\n",
    "    axes.flatten()[i].title.set_fontsize(10)\n",
    "if len(axes.flatten()) > len(cts_i):\n",
    "    for a in axes.flatten()[len(cts_i):]:\n",
    "        a.set_visible(False)\n",
    "fig.suptitle(\"Senescent Cell Spaceflight DEGs\", fontsize=\"xx-large\",\n",
    "             fontproperties=dict(family=\"serif\"))\n",
    "res_pathways_snc_top_vs_no[[\"Term Short\", \"Genes\"] + list(\n",
    "    res_pathways_snc_top_vs_no.columns.difference([\n",
    "        \"Term\", \"Genes\", \"Term Short\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfecf592",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_degs = out_edgr_old_sf_v_gc.copy()\n",
    "# df_degs = out_edgr_batches_snc_top.copy()\n",
    "# df_degs = res_snc_grps.copy()\n",
    "\n",
    "# factor_name = col_celltype\n",
    "# factor_name = col_condition\n",
    "factor_name = col_batch\n",
    "\n",
    "df_degs = res_rank_genes_top.loc[:, \"Overall\", :]\n",
    "df_degs = df_degs.assign(log_fc=df_degs.logfoldchanges.abs())\n",
    "df_degs = df_degs.assign(abs_log_fc=df_degs.logfoldchanges_abs)\n",
    "\n",
    "names = gp.get_library_name()\n",
    "# [i for i in names if \"brain\" in i.lower() or \"mouse\" in i.lower()]\n",
    "pathways_snc_vs_no, res_pathways_snc_vs_no = {}, {}\n",
    "thresh_p_path = 0.05\n",
    "# perts_geo = [\"Aging\", \"Drug\", \"Disease\"]\n",
    "perts_geo = [\"Aging\", \"Disease\"]\n",
    "other_sets = [\"GTEx_Aging_Signatures_2021\", f\"HDSigDB_{species}_2021\",\n",
    "              # \"DGIdb_Drug_Targets_2024\",\n",
    "              \"MSigDB_Hallmark_2020\",\n",
    "              \"MSigDB_Oncogenic_Signatures\", \"OMIM_Disease\",\n",
    "              # \"Reactome_Pathways_2024\", \"Reactome_2022\",\n",
    "              f\"WikiPathways_2024_{species}\", f\"WikiPathways_2019_{species}\"]\n",
    "other_sets += [i for i in names if \"GO_Biological_Process\" in i]\n",
    "# if species.lower() == \"mouse\":\n",
    "#     other_sets += [\"PerturbAtlas_MouseGenePerturbationSigs\"]\n",
    "for x in df_degs.reset_index()[factor_name].unique():\n",
    "    pathways_snc_vs_no[x], res_pathways_snc_vs_no[x] = {}, {}\n",
    "    for i in [\"up\", \"down\"]:\n",
    "        degs_tmp = df_degs.loc[x]\n",
    "        if i == \"up\":\n",
    "            degs_tmp = list(degs_tmp[degs_tmp.log_fc > 0].index.values)\n",
    "        else:\n",
    "            degs_tmp = list(degs_tmp[degs_tmp.log_fc < 0].index.values)\n",
    "        if len(degs_tmp) > 0:\n",
    "            g_s = [f\"{u}_Perturbations_from_GEO_{i}\" for u in perts_geo] + [\n",
    "                f\"Disease_Signatures_from_GEO_{i}_2014\"] + other_sets\n",
    "            g_s = [g for g in g_s if g in names]\n",
    "            pathways_snc_vs_no[x][i] = gp.enrichr(\n",
    "                gene_list=degs_tmp, gene_sets=g_s,\n",
    "                organism=species, cutoff=0.5)\n",
    "            res_pathways_snc_vs_no[x][i] = pathways_snc_vs_no[x][i].results[\n",
    "                pathways_snc_vs_no[x][i].results[\n",
    "                    \"Adjusted P-value\"] < thresh_p_path]\n",
    "    res_pathways_snc_vs_no[x] = pd.concat(res_pathways_snc_vs_no[x],\n",
    "                                          names=[\"Direction\"])\n",
    "res_pathways_snc_vs_no = pd.concat(res_pathways_snc_vs_no,\n",
    "                                   names=[col_celltype])\n",
    "res_pathways_snc_vs_no.loc[\n",
    "    :, \"Term Short\"] = res_pathways_snc_vs_no.Term.apply(\n",
    "        lambda x: re.sub(\"Expression Of \", \"\", re.sub(\".*.xls\", \"\", re.sub(\n",
    "            \".*.XLSX\", \"\",\n",
    "            re.sub(\".*Supplementary Table [0-9]+[-]?\", \"\", re.sub(\n",
    "                \".*.xlsx.\", \"\", re.sub(\".*Supplementary Data ..\", \"\", re.sub(\n",
    "                    \"PMID([0-9])+\", \"\", re.sub(\n",
    "                        \"([0-9]+) years v ([0-9]+) years\", \"\\\\1 v \\\\2 yrs.\",\n",
    "                        re.sub(\"aging:([0-9]+)\", \"aging\", re.sub(\n",
    "                            \"([0-9]+) months v ([0-9]+) mo(nths)?\",\n",
    "                            \"\\\\1 v \\\\2 mo.\", re.sub(\"SRP[0-9]+ \", \"\", re.sub(\n",
    "                                \"GSE[0-9]+ \", \"\", x)))))))))))))\n",
    "cts_i = res_pathways_snc_vs_no.reset_index()[col_celltype].unique()\n",
    "res_pathways_snc_top_vs_no = res_pathways_snc_vs_no.reset_index().groupby([\n",
    "    col_celltype, \"Direction\"]).apply(lambda x: x.sort_values(\n",
    "        \"Adjusted P-value\", ascending=True).head(10), include_groups=False)\n",
    "fig, axes = plt.subplots(*scflow.pl.square_grid(cts_i), figsize=(\n",
    "    20, 25), gridspec_kw=dict(wspace=1.5, top=0.95), squeeze=False)\n",
    "for i, x in enumerate(cts_i):\n",
    "    gp.dotplot(\n",
    "        res_pathways_snc_top_vs_no.loc[x].reset_index(),\n",
    "        column=\"Adjusted P-value\",\n",
    "        # column=\"Overlap\",\n",
    "        x=\"Direction\", y=\"Term Short\",\n",
    "        ax=axes.flatten()[i], size=20, top_term=100,\n",
    "        title=x, xticklabels_rot=45, yticklabels_rot=45,\n",
    "        show_ring=True, marker=\"o\", cutoff=0.5)\n",
    "    axes.flatten()[i].tick_params(axis=\"y\", labelsize=8,\n",
    "                                  labelfontfamily=\"serif\")\n",
    "    axes.flatten()[i].tick_params(axis=\"x\", labelsize=10,\n",
    "                                  labelfontfamily=\"serif\")\n",
    "    axes.flatten()[i].title.set_fontsize(10)\n",
    "if len(axes.flatten()) > len(cts_i):\n",
    "    for a in axes.flatten()[len(cts_i):]:\n",
    "        a.set_visible(False)\n",
    "fig.suptitle(\"Senescent Cell Spaceflight DEGs\",\n",
    "             fontsize=\"xx-large\",\n",
    "             fontproperties=dict(family=\"serif\"))\n",
    "res_pathways_snc_top_vs_no[[\"Term Short\", \"Genes\"] + list(\n",
    "    res_pathways_snc_top_vs_no.columns.difference([\n",
    "        \"Term\", \"Genes\", \"Term Short\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c26efd82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_degs = out_edgr_old_sf_v_gc.copy()\n",
    "# df_degs = out_edgr_batches_snc_top.copy()\n",
    "# df_degs = res_snc_grps.copy()\n",
    "\n",
    "factor_name = col_celltype\n",
    "# factor_name = col_condition\n",
    "\n",
    "iii = 2\n",
    "df_degs = res_rank_genes_top.loc[self.rna.obs[col_batch].unique()[iii]]\n",
    "df_degs = df_degs.assign(log_fc=df_degs.logfoldchanges.abs())\n",
    "df_degs = df_degs.assign(abs_log_fc=df_degs.logfoldchanges_abs)\n",
    "print(self.rna.obs[col_batch].unique()[iii])\n",
    "\n",
    "names = gp.get_library_name()\n",
    "# [i for i in names if \"brain\" in i.lower() or \"mouse\" in i.lower()]\n",
    "pathways_snc_vs_no, res_pathways_snc_vs_no = {}, {}\n",
    "thresh_p_path = 0.05\n",
    "# perts_geo = [\"Aging\", \"Drug\", \"Disease\"]\n",
    "perts_geo = [\"Aging\", \"Disease\"]\n",
    "other_sets = [\"GTEx_Aging_Signatures_2021\", f\"HDSigDB_{species}_2021\",\n",
    "              # \"DGIdb_Drug_Targets_2024\",\n",
    "              \"MSigDB_Hallmark_2020\",\n",
    "              \"MSigDB_Oncogenic_Signatures\", \"OMIM_Disease\",\n",
    "              # \"Reactome_Pathways_2024\", \"Reactome_2022\",\n",
    "              f\"WikiPathways_2024_{species}\", f\"WikiPathways_2019_{species}\"]\n",
    "other_sets += [i for i in names if \"GO_Biological_Process\" in i]\n",
    "# if species.lower() == \"mouse\":\n",
    "#     other_sets += [\"PerturbAtlas_MouseGenePerturbationSigs\"]\n",
    "for x in df_degs.reset_index()[factor_name].unique():\n",
    "    pathways_snc_vs_no[x], res_pathways_snc_vs_no[x] = {}, {}\n",
    "    for i in [\"up\", \"down\"]:\n",
    "        degs_tmp = df_degs.loc[x]\n",
    "        if i == \"up\":\n",
    "            degs_tmp = list(degs_tmp[degs_tmp.log_fc > 0].index.values)\n",
    "        else:\n",
    "            degs_tmp = list(degs_tmp[degs_tmp.log_fc < 0].index.values)\n",
    "        if len(degs_tmp) > 0:\n",
    "            g_s = [f\"{u}_Perturbations_from_GEO_{i}\" for u in perts_geo] + [\n",
    "                f\"Disease_Signatures_from_GEO_{i}_2014\"] + other_sets\n",
    "            g_s = [g for g in g_s if g in names]\n",
    "            pathways_snc_vs_no[x][i] = gp.enrichr(\n",
    "                gene_list=degs_tmp, gene_sets=g_s,\n",
    "                organism=species, cutoff=0.5)\n",
    "            res_pathways_snc_vs_no[x][i] = pathways_snc_vs_no[x][i].results[\n",
    "                pathways_snc_vs_no[x][i].results[\n",
    "                    \"Adjusted P-value\"] < thresh_p_path]\n",
    "    res_pathways_snc_vs_no[x] = pd.concat(res_pathways_snc_vs_no[x],\n",
    "                                          names=[\"Direction\"])\n",
    "res_pathways_snc_vs_no = pd.concat(res_pathways_snc_vs_no,\n",
    "                                   names=[col_celltype])\n",
    "res_pathways_snc_vs_no.loc[\n",
    "    :, \"Term Short\"] = res_pathways_snc_vs_no.Term.apply(\n",
    "        lambda x: re.sub(\"Expression Of \", \"\", re.sub(\".*.xls\", \"\", re.sub(\n",
    "            \".*.XLSX\", \"\",\n",
    "            re.sub(\".*Supplementary Table [0-9]+[-]?\", \"\", re.sub(\n",
    "                \".*.xlsx.\", \"\", re.sub(\".*Supplementary Data ..\", \"\", re.sub(\n",
    "                    \"PMID([0-9])+\", \"\", re.sub(\n",
    "                        \"([0-9]+) years v ([0-9]+) years\", \"\\\\1 v \\\\2 yrs.\",\n",
    "                        re.sub(\"aging:([0-9]+)\", \"aging\", re.sub(\n",
    "                            \"([0-9]+) months v ([0-9]+) mo(nths)?\",\n",
    "                            \"\\\\1 v \\\\2 mo.\", re.sub(\"SRP[0-9]+ \", \"\", re.sub(\n",
    "                                \"GSE[0-9]+ \", \"\", x)))))))))))))\n",
    "cts_i = res_pathways_snc_vs_no.reset_index()[col_celltype].unique()\n",
    "res_pathways_snc_top_vs_no = res_pathways_snc_vs_no.reset_index().groupby([\n",
    "    col_celltype, \"Direction\"]).apply(lambda x: x.sort_values(\n",
    "        \"Adjusted P-value\", ascending=True).head(10), include_groups=False)\n",
    "fig, axes = plt.subplots(*scflow.pl.square_grid(cts_i), figsize=(\n",
    "    20, 25), gridspec_kw=dict(wspace=1.5, top=0.95), squeeze=False)\n",
    "for i, x in enumerate(cts_i):\n",
    "    gp.dotplot(\n",
    "        res_pathways_snc_top_vs_no.loc[x].reset_index(),\n",
    "        column=\"Adjusted P-value\",\n",
    "        # column=\"Overlap\",\n",
    "        x=\"Direction\", y=\"Term Short\",\n",
    "        ax=axes.flatten()[i], size=20, top_term=100,\n",
    "        title=x, xticklabels_rot=45, yticklabels_rot=45,\n",
    "        show_ring=True, marker=\"o\", cutoff=0.5)\n",
    "    axes.flatten()[i].tick_params(axis=\"y\", labelsize=8,\n",
    "                                  labelfontfamily=\"serif\")\n",
    "    axes.flatten()[i].tick_params(axis=\"x\", labelsize=10,\n",
    "                                  labelfontfamily=\"serif\")\n",
    "    axes.flatten()[i].title.set_fontsize(10)\n",
    "if len(axes.flatten()) > len(cts_i):\n",
    "    for a in axes.flatten()[len(cts_i):]:\n",
    "        a.set_visible(False)\n",
    "fig.suptitle(\"Senescent Cell Spaceflight DEGs\", fontsize=\"xx-large\",\n",
    "             fontproperties=dict(family=\"serif\"))\n",
    "res_pathways_snc_top_vs_no[[\"Term Short\", \"Genes\"] + list(\n",
    "    res_pathways_snc_top_vs_no.columns.difference([\n",
    "        \"Term\", \"Genes\", \"Term Short\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9cc30a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_degs = out_edgr_old_sf_v_gc.copy()\n",
    "# df_degs = out_edgr_batches_snc_top.copy()\n",
    "# df_degs = res_snc_grps.copy()\n",
    "\n",
    "factor_name = col_celltype\n",
    "# factor_name = col_condition\n",
    "\n",
    "iii = 1\n",
    "df_degs = res_rank_genes_top.loc[self.rna.obs[col_batch].unique()[iii]]\n",
    "df_degs = df_degs.assign(log_fc=df_degs.logfoldchanges.abs())\n",
    "df_degs = df_degs.assign(abs_log_fc=df_degs.logfoldchanges_abs)\n",
    "print(self.rna.obs[col_batch].unique()[iii])\n",
    "\n",
    "names = gp.get_library_name()\n",
    "# [i for i in names if \"brain\" in i.lower() or \"mouse\" in i.lower()]\n",
    "pathways_snc_vs_no, res_pathways_snc_vs_no = {}, {}\n",
    "thresh_p_path = 0.05\n",
    "# perts_geo = [\"Aging\", \"Drug\", \"Disease\"]\n",
    "perts_geo = [\"Aging\", \"Disease\"]\n",
    "other_sets = [\"GTEx_Aging_Signatures_2021\", f\"HDSigDB_{species}_2021\",\n",
    "              # \"DGIdb_Drug_Targets_2024\",\n",
    "              \"MSigDB_Hallmark_2020\",\n",
    "              \"MSigDB_Oncogenic_Signatures\", \"OMIM_Disease\",\n",
    "              # \"Reactome_Pathways_2024\", \"Reactome_2022\",\n",
    "              f\"WikiPathways_2024_{species}\", f\"WikiPathways_2019_{species}\"]\n",
    "other_sets += [i for i in names if \"GO_Biological_Process\" in i]\n",
    "# if species.lower() == \"mouse\":\n",
    "#     other_sets += [\"PerturbAtlas_MouseGenePerturbationSigs\"]\n",
    "for x in df_degs.reset_index()[factor_name].unique():\n",
    "    pathways_snc_vs_no[x], res_pathways_snc_vs_no[x] = {}, {}\n",
    "    for i in [\"up\", \"down\"]:\n",
    "        degs_tmp = df_degs.loc[x]\n",
    "        if i == \"up\":\n",
    "            degs_tmp = list(degs_tmp[degs_tmp.log_fc > 0].index.values)\n",
    "        else:\n",
    "            degs_tmp = list(degs_tmp[degs_tmp.log_fc < 0].index.values)\n",
    "        if len(degs_tmp) > 0:\n",
    "            g_s = [f\"{u}_Perturbations_from_GEO_{i}\" for u in perts_geo] + [\n",
    "                f\"Disease_Signatures_from_GEO_{i}_2014\"] + other_sets\n",
    "            g_s = [g for g in g_s if g in names]\n",
    "            pathways_snc_vs_no[x][i] = gp.enrichr(\n",
    "                gene_list=degs_tmp, gene_sets=g_s,\n",
    "                organism=species, cutoff=0.5)\n",
    "            res_pathways_snc_vs_no[x][i] = pathways_snc_vs_no[x][i].results[\n",
    "                pathways_snc_vs_no[x][i].results[\n",
    "                    \"Adjusted P-value\"] < thresh_p_path]\n",
    "    res_pathways_snc_vs_no[x] = pd.concat(res_pathways_snc_vs_no[x],\n",
    "                                          names=[\"Direction\"])\n",
    "res_pathways_snc_vs_no = pd.concat(res_pathways_snc_vs_no,\n",
    "                                   names=[col_celltype])\n",
    "res_pathways_snc_vs_no.loc[\n",
    "    :, \"Term Short\"] = res_pathways_snc_vs_no.Term.apply(\n",
    "        lambda x: re.sub(\"Expression Of \", \"\", re.sub(\".*.xls\", \"\", re.sub(\n",
    "            \".*.XLSX\", \"\",\n",
    "            re.sub(\".*Supplementary Table [0-9]+[-]?\", \"\", re.sub(\n",
    "                \".*.xlsx.\", \"\", re.sub(\".*Supplementary Data ..\", \"\", re.sub(\n",
    "                    \"PMID([0-9])+\", \"\", re.sub(\n",
    "                        \"([0-9]+) years v ([0-9]+) years\", \"\\\\1 v \\\\2 yrs.\",\n",
    "                        re.sub(\"aging:([0-9]+)\", \"aging\", re.sub(\n",
    "                            \"([0-9]+) months v ([0-9]+) mo(nths)?\",\n",
    "                            \"\\\\1 v \\\\2 mo.\", re.sub(\"SRP[0-9]+ \", \"\", re.sub(\n",
    "                                \"GSE[0-9]+ \", \"\", x)))))))))))))\n",
    "cts_i = res_pathways_snc_vs_no.reset_index()[col_celltype].unique()\n",
    "res_pathways_snc_top_vs_no = res_pathways_snc_vs_no.reset_index().groupby([\n",
    "    col_celltype, \"Direction\"]).apply(lambda x: x.sort_values(\n",
    "        \"Adjusted P-value\", ascending=True).head(10), include_groups=False)\n",
    "fig, axes = plt.subplots(*scflow.pl.square_grid(cts_i), figsize=(\n",
    "    20, 25), gridspec_kw=dict(wspace=1.5, top=0.95), squeeze=False)\n",
    "for i, x in enumerate(cts_i):\n",
    "    gp.dotplot(\n",
    "        res_pathways_snc_top_vs_no.loc[x].reset_index(),\n",
    "        column=\"Adjusted P-value\",\n",
    "        # column=\"Overlap\",\n",
    "        x=\"Direction\", y=\"Term Short\",\n",
    "        ax=axes.flatten()[i], size=20, top_term=100,\n",
    "        title=x, xticklabels_rot=45, yticklabels_rot=45,\n",
    "        show_ring=True, marker=\"o\", cutoff=0.5)\n",
    "    axes.flatten()[i].tick_params(axis=\"y\", labelsize=8,\n",
    "                                  labelfontfamily=\"serif\")\n",
    "    axes.flatten()[i].tick_params(axis=\"x\", labelsize=10,\n",
    "                                  labelfontfamily=\"serif\")\n",
    "    axes.flatten()[i].title.set_fontsize(10)\n",
    "if len(axes.flatten()) > len(cts_i):\n",
    "    for a in axes.flatten()[len(cts_i):]:\n",
    "        a.set_visible(False)\n",
    "fig.suptitle(\"Senescent Cell Spaceflight DEGs\", fontsize=\"xx-large\",\n",
    "             fontproperties=dict(family=\"serif\"))\n",
    "res_pathways_snc_top_vs_no[[\"Term Short\", \"Genes\"] + list(\n",
    "    res_pathways_snc_top_vs_no.columns.difference([\n",
    "        \"Term\", \"Genes\", \"Term Short\"]))]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rsc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
